--- a/net/minecraft/client/renderer/BlockFluidRenderer.java
+++ b/net/minecraft/client/renderer/BlockFluidRenderer.java
@@ -18,7 +18,6 @@
     private TextureAtlasSprite[] field_178272_a = new TextureAtlasSprite[2];
     private TextureAtlasSprite[] field_178271_b = new TextureAtlasSprite[2];
     private static final String __OBFID = "CL_00002519";
-    private static final String __OBFID = "CL_00002519";
 
     public BlockFluidRenderer()
     {
