--- a/net/minecraft/client/renderer/EntityRenderer.java
+++ b/net/minecraft/client/renderer/EntityRenderer.java
@@ -241,7 +241,6 @@
     private boolean lastShowDebugInfo = false;
     private boolean showExtendedDebugInfo = false;
     private long lastErrorCheckTimeMs = 0L;
-    private static final String __OBFID = "CL_00000947";
 
     public EntityRenderer(Minecraft p_i45076_1_, IResourceManager p_i45076_2_)
     {
@@ -1390,41 +1389,26 @@
                     var11.addCrashSectionCallable("Screen name", new Callable()
                     {
                         private static final String __OBFID = "CL_00000948";
-                        private static final String __OBFID = "CL_00000948";
                         public String call()
                         {
                             return EntityRenderer.this.mc.currentScreen.getClass().getCanonicalName();
                         }
-                        public Object call() throws Exception
-                        {
-                            return this.call();
-                        }
                     });
                     var11.addCrashSectionCallable("Mouse location", new Callable()
                     {
                         private static final String __OBFID = "CL_00000950";
-                        private static final String __OBFID = "CL_00000950";
                         public String call()
                         {
                             return String.format("Scaled: (%d, %d). Absolute: (%d, %d)", new Object[] {Integer.valueOf(var161), Integer.valueOf(var171), Integer.valueOf(Mouse.getX()), Integer.valueOf(Mouse.getY())});
                         }
-                        public Object call() throws Exception
-                        {
-                            return this.call();
-                        }
                     });
                     var11.addCrashSectionCallable("Screen size", new Callable()
                     {
                         private static final String __OBFID = "CL_00000951";
-                        private static final String __OBFID = "CL_00000951";
                         public String call()
                         {
                             return String.format("Scaled: (%d, %d). Absolute: (%d, %d). Scale factor of %d", new Object[] {Integer.valueOf(var133.getScaledWidth()), Integer.valueOf(var133.getScaledHeight()), Integer.valueOf(EntityRenderer.this.mc.displayWidth), Integer.valueOf(EntityRenderer.this.mc.displayHeight), Integer.valueOf(var133.getScaleFactor())});
                         }
-                        public Object call() throws Exception
-                        {
-                            return this.call();
-                        }
                     });
                     throw new ReportedException(var10);
                 }
