--- a/net/minecraft/client/renderer/WorldRenderer.java
+++ b/net/minecraft/client/renderer/WorldRenderer.java
@@ -46,7 +46,6 @@
     private boolean[] drawnIcons = new boolean[256];
     private TextureAtlasSprite[] quadSprites = null;
     private TextureAtlasSprite quadSprite = null;
-    private static final String __OBFID = "CL_00000942";
 
     public WorldRenderer(int p_i46275_1_)
     {
@@ -773,7 +772,6 @@
         private final VertexFormat field_179018_e;
         private static final String __OBFID = "CL_00002568";
         public TextureAtlasSprite[] stateQuadSprites;
-        private static final String __OBFID = "CL_00002568";
 
         public State(int[] p_i46274_2_, int p_i46274_3_, int p_i46274_4_, VertexFormat p_i46274_5_, TextureAtlasSprite[] sprites)
         {
@@ -809,7 +807,6 @@
     {
         static final int[] field_178959_a = new int[VertexFormatElement.EnumUseage.values().length];
         private static final String __OBFID = "CL_00002569";
-        private static final String __OBFID = "CL_00002569";
 
         static
         {
