diff -u a/fffix.py b/fffix.py
--- a/fffix.py
+++ b/fffix.py
@@ -194,13 +194,15 @@
         args = ''
         for arg in match.group('arguments').split(', '):
             args += '%s p_%s_%s_, ' % (arg.split(' ')[0], match.group('number'), arg.split(' ')[1][3:])
         args = args[:-2]
         
         return match.group(0).replace(match.group('arguments'), args)
-    buf = _REGEXP['abstract'].sub(abstract_match, buf)
+    
+    if src_file.find("ChunkRenderDispatcher") == -1:
+        buf = _REGEXP['abstract'].sub(abstract_match, buf)
 
     with open(tmp_file, 'w') as fh:
         fh.write(buf)
     shutil.move(tmp_file, src_file)
 
 
