--- a/net/minecraft/world/gen/structure/MapGenStructure.java
+++ b/net/minecraft/world/gen/structure/MapGenStructure.java
@@ -10,6 +10,7 @@
 import net.minecraft.crash.CrashReportCategory;
 import net.minecraft.nbt.NBTBase;
 import net.minecraft.nbt.NBTTagCompound;
+import net.minecraft.src.Reflector;
 import net.minecraft.util.BlockPos;
 import net.minecraft.util.LongHashMap;
 import net.minecraft.util.ReportedException;
@@ -17,6 +18,7 @@
 import net.minecraft.world.World;
 import net.minecraft.world.chunk.ChunkPrimer;
 import net.minecraft.world.gen.MapGenBase;
+import net.minecraft.world.storage.MapStorage;
 
 public abstract class MapGenStructure extends MapGenBase
 {
@@ -30,7 +32,6 @@
     protected Map structureMap = Maps.newHashMap();
     private static final String __OBFID = "CL_00000505";
     private LongHashMap structureLongMap = new LongHashMap();
-    private static final String __OBFID = "CL_00000505";
 
     public abstract String getStructureName();
 
@@ -59,42 +60,27 @@
                 var9.addCrashSectionCallable("Is feature chunk", new Callable()
                 {
                     private static final String __OBFID = "CL_00000506";
-                    private static final String __OBFID = "CL_00000506";
                     public String call()
                     {
                         return MapGenStructure.this.canSpawnStructureAtCoords(p_180701_2_, p_180701_3_) ? "True" : "False";
                     }
-                    public Object call() throws Exception
-                    {
-                        return this.call();
-                    }
                 });
                 var9.addCrashSection("Chunk location", String.format("%d,%d", new Object[] {Integer.valueOf(p_180701_2_), Integer.valueOf(p_180701_3_)}));
                 var9.addCrashSectionCallable("Chunk pos hash", new Callable()
                 {
                     private static final String __OBFID = "CL_00000507";
-                    private static final String __OBFID = "CL_00000507";
                     public String call()
                     {
                         return String.valueOf(ChunkCoordIntPair.chunkXZ2Int(p_180701_2_, p_180701_3_));
                     }
-                    public Object call() throws Exception
-                    {
-                        return this.call();
-                    }
                 });
                 var9.addCrashSectionCallable("Structure type", new Callable()
                 {
                     private static final String __OBFID = "CL_00000508";
-                    private static final String __OBFID = "CL_00000508";
                     public String call()
                     {
                         return MapGenStructure.this.getClass().getCanonicalName();
                     }
-                    public Object call() throws Exception
-                    {
-                        return this.call();
-                    }
                 });
                 throw new ReportedException(var8);
             }
@@ -260,22 +246,41 @@
     {
         if (this.field_143029_e == null)
         {
-            this.field_143029_e = (MapGenStructureData)worldIn.loadItemData(MapGenStructureData.class, this.getStructureName());
+            MapStorage var2;
+
+            if (Reflector.ForgeWorld_getPerWorldStorage.exists())
+            {
+                var2 = (MapStorage)Reflector.call(worldIn, Reflector.ForgeWorld_getPerWorldStorage, new Object[0]);
+                this.field_143029_e = (MapGenStructureData)var2.loadData(MapGenStructureData.class, this.getStructureName());
+            }
+            else
+            {
+                this.field_143029_e = (MapGenStructureData)worldIn.loadItemData(MapGenStructureData.class, this.getStructureName());
+            }
 
             if (this.field_143029_e == null)
             {
                 this.field_143029_e = new MapGenStructureData(this.getStructureName());
-                worldIn.setItemData(this.getStructureName(), this.field_143029_e);
+
+                if (Reflector.ForgeWorld_getPerWorldStorage.exists())
+                {
+                    var2 = (MapStorage)Reflector.call(worldIn, Reflector.ForgeWorld_getPerWorldStorage, new Object[0]);
+                    var2.setData(this.getStructureName(), this.field_143029_e);
+                }
+                else
+                {
+                    worldIn.setItemData(this.getStructureName(), this.field_143029_e);
+                }
             }
             else
             {
-                NBTTagCompound var2 = this.field_143029_e.func_143041_a();
-                Iterator var3 = var2.getKeySet().iterator();
+                NBTTagCompound var21 = this.field_143029_e.func_143041_a();
+                Iterator var3 = var21.getKeySet().iterator();
 
                 while (var3.hasNext())
                 {
                     String var4 = (String)var3.next();
-                    NBTBase var5 = var2.getTag(var4);
+                    NBTBase var5 = var21.getTag(var4);
 
                     if (var5.getId() == 10)
                     {
