--- a/net/minecraft/world/chunk/storage/ExtendedBlockStorage.java
+++ b/net/minecraft/world/chunk/storage/ExtendedBlockStorage.java
@@ -31,7 +31,6 @@
     /** The NibbleArray containing a block of Sky-light data. */
     private NibbleArray skylightArray;
     private static final String __OBFID = "CL_00000375";
-    private static final String __OBFID = "CL_00000375";
 
     public ExtendedBlockStorage(int y, boolean storeSkylight)
     {
