--- a/net/minecraft/client/shader/Framebuffer.java
+++ b/net/minecraft/client/shader/Framebuffer.java
@@ -6,7 +6,13 @@
 import net.minecraft.client.renderer.Tessellator;
 import net.minecraft.client.renderer.WorldRenderer;
 import net.minecraft.client.renderer.texture.TextureUtil;
+import org.lwjgl.opengl.ARBFramebufferObject;
+import org.lwjgl.opengl.ContextCapabilities;
+import org.lwjgl.opengl.EXTFramebufferObject;
 import org.lwjgl.opengl.GL11;
+import org.lwjgl.opengl.GL30;
+import org.lwjgl.opengl.GL32;
+import org.lwjgl.opengl.GLContext;
 
 public class Framebuffer
 {
@@ -22,8 +28,19 @@
     public int framebufferFilter;
     private static final String __OBFID = "CL_00000959";
 
+    /** MINECRIFT */
+    public boolean genMipMaps = false;
+	/** END MINECRIFT */
+
     public Framebuffer(int p_i45078_1_, int p_i45078_2_, boolean p_i45078_3_)
     {
+	/** MINECRIFT */
+        this(p_i45078_1_, p_i45078_2_, p_i45078_3_, false);
+    }
+
+    public Framebuffer(int p_i45078_1_, int p_i45078_2_, boolean p_i45078_3_, boolean generateMipMaps)
+    {
+	/** END MINECRIFT */
         this.useDepth = p_i45078_3_;
         this.framebufferObject = -1;
         this.framebufferTexture = -1;
@@ -33,6 +50,10 @@
         this.framebufferColor[1] = 1.0F;
         this.framebufferColor[2] = 1.0F;
         this.framebufferColor[3] = 0.0F;
+
+        /** MINECRIFT */
+        this.genMipMaps = generateMipMaps;
+		/** END MINECRIFT */
         this.createBindFramebuffer(p_i45078_1_, p_i45078_2_);
     }
 
@@ -53,7 +74,9 @@
             }
 
             this.createFramebuffer(p_147613_1_, p_147613_2_);
-            this.checkFramebufferComplete();
+			/** MINECRIFT */
+			//this.checkFramebufferComplete();
+			/** END MINECRIFT */
             OpenGlHelper.func_153171_g(OpenGlHelper.field_153198_e, 0);
         }
     }
@@ -110,6 +133,10 @@
             this.setFramebufferFilter(9728);
             GlStateManager.func_179144_i(this.framebufferTexture);
             GL11.glTexImage2D(GL11.GL_TEXTURE_2D, 0, GL11.GL_RGBA8, this.framebufferTextureWidth, this.framebufferTextureHeight, 0, GL11.GL_RGBA, GL11.GL_UNSIGNED_BYTE, (ByteBuffer)null);
+            /** MINECRIFT */
+            if (this.genMipMaps)
+                genMipMaps(); // TODO: Check GLContext capabilities
+            /** END MINECRIFT */            
             OpenGlHelper.func_153171_g(OpenGlHelper.field_153198_e, this.framebufferObject);
             OpenGlHelper.func_153188_a(OpenGlHelper.field_153198_e, OpenGlHelper.field_153200_g, 3553, this.framebufferTexture, 0);
 
@@ -120,18 +147,28 @@
                 OpenGlHelper.func_153190_b(OpenGlHelper.field_153198_e, OpenGlHelper.field_153201_h, OpenGlHelper.field_153199_f, this.depthBuffer);
             }
 
+            /** MINECRIFT */
+            this.checkFramebufferComplete();
+            /** END MINECRIFT */
             this.framebufferClear();
             this.unbindFramebufferTexture();
         }
     }
 
+	/** MINECRIFT */
+    public void genMipMaps()
+    {
+        GL30.glGenerateMipmap(GL11.GL_TEXTURE_2D);
+    }
+	/** END MINECRIFT */
+
     public void setFramebufferFilter(int p_147607_1_)
     {
         if (OpenGlHelper.isFramebufferEnabled())
         {
             this.framebufferFilter = p_147607_1_;
             GlStateManager.func_179144_i(this.framebufferTexture);
-            GL11.glTexParameterf(GL11.GL_TEXTURE_2D, GL11.GL_TEXTURE_MIN_FILTER, (float)p_147607_1_);
+            GL11.glTexParameterf(GL11.GL_TEXTURE_2D, GL11.GL_TEXTURE_MIN_FILTER, (float) p_147607_1_);
             GL11.glTexParameterf(GL11.GL_TEXTURE_2D, GL11.GL_TEXTURE_MAG_FILTER, (float)p_147607_1_);
             GL11.glTexParameterf(GL11.GL_TEXTURE_2D, GL11.GL_TEXTURE_WRAP_S, 10496.0F);
             GL11.glTexParameterf(GL11.GL_TEXTURE_2D, GL11.GL_TEXTURE_WRAP_T, 10496.0F);
@@ -213,13 +250,20 @@
         this.framebufferColor[3] = p_147604_4_;
     }
 
+    /** MINECRIFT */
     public void framebufferRender(int p_147615_1_, int p_147615_2_)
     {
-        this.func_178038_a(p_147615_1_, p_147615_2_, true);
+        this.func_178038_a(p_147615_1_, p_147615_2_, true, 0, 0);
     }
 
     public void func_178038_a(int p_178038_1_, int p_178038_2_, boolean p_178038_3_)
     {
+        this.func_178038_a(p_178038_1_, p_178038_2_, p_178038_3_, 0, 0);
+    }
+
+    public void func_178038_a(int p_178038_1_, int p_178038_2_, boolean p_178038_3_, int left, int top)
+    {
+    /** END MINECRIFT */
         if (OpenGlHelper.isFramebufferEnabled())
         {
             GlStateManager.colorMask(true, true, true, false);
@@ -231,7 +275,9 @@
             GlStateManager.matrixMode(5888);
             GlStateManager.loadIdentity();
             GlStateManager.translate(0.0F, 0.0F, -2000.0F);
-            GlStateManager.viewport(0, 0, p_178038_1_, p_178038_2_);
+			/** MINECRIFT */
+            GlStateManager.viewport(left, top, p_178038_1_, p_178038_2_);
+			/** END MINECRIFT */            
             GlStateManager.func_179098_w();
             GlStateManager.disableLighting();
             GlStateManager.disableAlpha();
