--- a/net/minecraft/entity/projectile/EntityFireball.java
+++ b/net/minecraft/entity/projectile/EntityFireball.java
@@ -1,9 +1,15 @@
 package net.minecraft.entity.projectile;
 
 import java.util.List;
+
+import de.fruitfly.ovr.structs.Matrix4f;
+import de.fruitfly.ovr.structs.Vector3f;
 import net.minecraft.block.Block;
+import net.minecraft.client.Minecraft;
+import net.minecraft.client.entity.EntityPlayerSP;
 import net.minecraft.entity.Entity;
 import net.minecraft.entity.EntityLivingBase;
+import net.minecraft.entity.player.EntityPlayer;
 import net.minecraft.nbt.NBTTagCompound;
 import net.minecraft.nbt.NBTTagList;
 import net.minecraft.util.AxisAlignedBB;
@@ -59,12 +65,35 @@
         this.accelerationZ = p_i1760_12_ / var14 * 0.1D;
     }
 
-    public EntityFireball(World p_i1761_1_, EntityLivingBase p_i1761_2_, double p_i1761_3_, double p_i1761_5_, double p_i1761_7_)
+    public EntityFireball(World p_i1761_1_, EntityLivingBase shooter, double p_i1761_3_, double p_i1761_5_, double p_i1761_7_)
     {
         super(p_i1761_1_);
-        this.shootingEntity = p_i1761_2_;
+        this.shootingEntity = shooter;
         this.setSize(1.0F, 1.0F);
-        this.setLocationAndAngles(p_i1761_2_.posX, p_i1761_2_.posY, p_i1761_2_.posZ, p_i1761_2_.rotationYaw, p_i1761_2_.rotationPitch);
+        
+   
+        //JRBUDDA VIVE
+        Vec3 spawnPosition = shooter.getPosition(1);
+        
+        float yaw = shooter.rotationYaw;
+        float pitch = shooter.rotationPitch;
+     
+       if (shooter instanceof EntityPlayer)
+        {
+       	
+           Vec3 aimSource = Minecraft.getMinecraft().lookaimController.getAimSource(0);
+           Vec3 aimRotation = Minecraft.getMinecraft().lookaimController.getAimVector(0);
+           spawnPosition.xCoord = aimSource.xCoord + aimRotation.xCoord * 0.28f;
+           spawnPosition.yCoord = aimSource.yCoord - aimRotation.yCoord * 0.28f;
+           spawnPosition.zCoord = aimSource.zCoord + aimRotation.zCoord * 0.28f;
+     	
+           yaw = Minecraft.getMinecraft().lookaimController.getAimYaw();
+           pitch = Minecraft.getMinecraft().lookaimController.getAimPitch();
+        }
+    
+        this.setLocationAndAngles(spawnPosition.xCoord, spawnPosition.yCoord ,spawnPosition.zCoord, yaw, pitch);
+        //END VIVE
+        
         this.setPosition(this.posX, this.posY, this.posZ);
         this.yOffset = 0.0F;
         this.motionX = this.motionY = this.motionZ = 0.0D;
