--- a/net/minecraft/entity/projectile/EntityThrowable.java
+++ b/net/minecraft/entity/projectile/EntityThrowable.java
@@ -1,7 +1,12 @@
 package net.minecraft.entity.projectile;
 
 import java.util.List;
+
+import de.fruitfly.ovr.structs.Matrix4f;
+import de.fruitfly.ovr.structs.Vector3f;
 import net.minecraft.block.Block;
+import net.minecraft.client.Minecraft;
+import net.minecraft.client.entity.EntityPlayerSP;
 import net.minecraft.entity.Entity;
 import net.minecraft.entity.EntityLivingBase;
 import net.minecraft.entity.IProjectile;
@@ -49,15 +54,40 @@
         return distance < var3 * var3;
     }
 
-    public EntityThrowable(World p_i1777_1_, EntityLivingBase p_i1777_2_)
+    public EntityThrowable(World p_i1777_1_, EntityLivingBase shooter)
     {
         super(p_i1777_1_);
-        this.thrower = p_i1777_2_;
+        this.thrower = shooter;
         this.setSize(0.25F, 0.25F);
-        this.setLocationAndAngles(p_i1777_2_.posX, p_i1777_2_.posY + (double)p_i1777_2_.getEyeHeight(), p_i1777_2_.posZ, p_i1777_2_.rotationYaw, p_i1777_2_.rotationPitch);
-        this.posX -= (double)(MathHelper.cos(this.rotationYaw / 180.0F * (float)Math.PI) * 0.16F);
-        this.posY -= 0.10000000149011612D;
-        this.posZ -= (double)(MathHelper.sin(this.rotationYaw / 180.0F * (float)Math.PI) * 0.16F);
+        
+  
+        //JRBUDDA VIVE
+        Vec3 spawnPosition = shooter.getPosition(1);
+        spawnPosition.yCoord += shooter.getEyeHeight();
+        
+        float yaw = shooter.rotationYaw;
+        float pitch = shooter.rotationPitch;
+     
+     
+        if (shooter instanceof EntityPlayer)
+        { 	
+            Vec3 aimSource = Minecraft.getMinecraft().roomScale.getControllerPos_World(0);
+            Vec3 controllerDirection = Minecraft.getMinecraft().roomScale.getControllerMainDir_World();
+            spawnPosition.xCoord = aimSource.xCoord + controllerDirection.xCoord * 0.28f;
+            spawnPosition.yCoord = aimSource.yCoord - controllerDirection.yCoord * 0.28f;
+            spawnPosition.zCoord = aimSource.zCoord + controllerDirection.zCoord * 0.28f;
+      	
+            yaw = Minecraft.getMinecraft().roomScale.getControllerMainYaw_World();
+            pitch =Minecraft.getMinecraft().roomScale.getControllerMainPitch_World();
+        }
+    
+        this.setLocationAndAngles(spawnPosition.xCoord, spawnPosition.yCoord ,spawnPosition.zCoord, yaw, pitch);
+        //END VIVE
+        
+
+//        this.posX -= (double)(MathHelper.cos(this.rotationYaw / 180.0F * (float)Math.PI) * 0.16F);
+//        this.posY -= 0.10000000149011612D;
+//        this.posZ -= (double)(MathHelper.sin(this.rotationYaw / 180.0F * (float)Math.PI) * 0.16F);
         this.setPosition(this.posX, this.posY, this.posZ);
         this.yOffset = 0.0F;
         float var3 = 0.4F;
