--- a/net/minecraft/client/particle/EffectRenderer.java
+++ b/net/minecraft/client/particle/EffectRenderer.java
@@ -45,7 +45,6 @@
     private Random rand = new Random();
     private Map field_178932_g = Maps.newHashMap();
     private static final String __OBFID = "CL_00000915";
-    private static final String __OBFID = "CL_00000915";
 
     public EffectRenderer(World worldIn, TextureManager p_i1220_2_)
     {
@@ -221,28 +220,18 @@
             var4.addCrashSectionCallable("Particle", new Callable()
             {
                 private static final String __OBFID = "CL_00000916";
-                private static final String __OBFID = "CL_00000916";
                 public String call()
                 {
                     return p_178923_1_.toString();
                 }
-                public Object call() throws Exception
-                {
-                    return this.call();
-                }
             });
             var4.addCrashSectionCallable("Particle Type", new Callable()
             {
                 private static final String __OBFID = "CL_00000917";
-                private static final String __OBFID = "CL_00000917";
                 public String call()
                 {
                     return var5 == 0 ? "MISC_TEXTURE" : (var5 == 1 ? "TERRAIN_TEXTURE" : (var5 == 3 ? "ENTITY_PARTICLE_TEXTURE" : "Unknown - " + var5));
                 }
-                public Object call() throws Exception
-                {
-                    return this.call();
-                }
             });
             throw new ReportedException(var3);
         }
@@ -315,28 +304,18 @@
                             var16.addCrashSectionCallable("Particle", new Callable()
                             {
                                 private static final String __OBFID = "CL_00000918";
-                                private static final String __OBFID = "CL_00000918";
                                 public String call()
                                 {
                                     return var13.toString();
                                 }
-                                public Object call() throws Exception
-                                {
-                                    return this.call();
-                                }
                             });
                             var16.addCrashSectionCallable("Particle Type", new Callable()
                             {
                                 private static final String __OBFID = "CL_00000919";
-                                private static final String __OBFID = "CL_00000919";
                                 public String call()
                                 {
                                     return var8 == 0 ? "MISC_TEXTURE" : (var8 == 1 ? "TERRAIN_TEXTURE" : (var8 == 3 ? "ENTITY_PARTICLE_TEXTURE" : "Unknown - " + var8));
                                 }
-                                public Object call() throws Exception
-                                {
-                                    return this.call();
-                                }
                             });
                             throw new ReportedException(var15);
                         }
