--- a/net/minecraft/client/gui/achievement/GuiAchievements.java
+++ b/net/minecraft/client/gui/achievement/GuiAchievements.java
@@ -1,6 +1,8 @@
 package net.minecraft.client.gui.achievement;
 
 import java.util.Random;
+
+import com.mtbs3d.minecrift.control.GuiScreenNavigator;
 import net.minecraft.client.Minecraft;
 import net.minecraft.client.gui.GuiButton;
 import net.minecraft.client.gui.GuiOptionButton;
@@ -108,8 +110,9 @@
         else
         {
             int var4;
-
-            if (Mouse.isButtonDown(0))
+            /** MINECRIFT **/
+            if (Mouse.isButtonDown(0) || GuiScreenNavigator.selectDepressed)
+            /** END MINECRIFT **/
             {
                 var4 = (this.width - this.field_146555_f) / 2;
                 int var5 = (this.height - this.field_146557_g) / 2;
