--- a/net/minecraft/client/renderer/texture/Stitcher.java
+++ b/net/minecraft/client/renderer/texture/Stitcher.java
@@ -24,7 +24,6 @@
     /** Max size (width or height) of a single tile */
     private final int maxTileDimension;
     private static final String __OBFID = "CL_00001054";
-    private static final String __OBFID = "CL_00001054";
 
     public Stitcher(int p_i45095_1_, int p_i45095_2_, boolean p_i45095_3_, int p_i45095_4_, int p_i45095_5_)
     {
@@ -233,7 +232,6 @@
         private boolean rotated;
         private float scaleFactor = 1.0F;
         private static final String __OBFID = "CL_00001055";
-        private static final String __OBFID = "CL_00001055";
 
         public Holder(TextureAtlasSprite p_i45094_1_, int p_i45094_2_)
         {
@@ -323,7 +321,6 @@
         private List subSlots;
         private Stitcher.Holder holder;
         private static final String __OBFID = "CL_00001056";
-        private static final String __OBFID = "CL_00001056";
 
         public Slot(int par1, int par2, int par3, int par4)
         {
