--- a/net/minecraft/client/renderer/texture/TextureAtlasSprite.java
+++ b/net/minecraft/client/renderer/texture/TextureAtlasSprite.java
@@ -44,7 +44,6 @@
     public int sheetHeight;
     public int glSpriteTextureId = -1;
     public TextureAtlasSprite spriteSingle = null;
-    private static final String __OBFID = "CL_00001062";
 
     private TextureAtlasSprite(TextureAtlasSprite parent)
     {
@@ -419,7 +418,6 @@
                     var7.addCrashSectionCallable("Frame sizes", new Callable()
                     {
                         private static final String __OBFID = "CL_00001063";
-                        private static final String __OBFID = "CL_00001063";
                         public String call()
                         {
                             StringBuilder var1 = new StringBuilder();
@@ -440,10 +438,6 @@
 
                             return var1.toString();
                         }
-                        public Object call() throws Exception
-                        {
-                            return this.call();
-                        }
                     });
                     throw new ReportedException(var6);
                 }
