--- a/net/minecraft/src/BetterGrass.java
+++ b/net/minecraft/src/BetterGrass.java
@@ -32,9 +32,9 @@
     {
         if (facing != EnumFacing.UP && facing != EnumFacing.DOWN)
         {
-            if (block instanceof BlockMycelium && Config.isBetterGrassFancy())
+            if (block instanceof BlockMycelium)
             {
-                return getBlockAt(blockPos, -1, facing, blockAccess) == Blocks.mycelium ? modelCubeMycelium.func_177551_a(facing) : quads;
+                return Config.isBetterGrassFancy() ? (getBlockAt(blockPos.offsetDown(), facing, blockAccess) == Blocks.mycelium ? modelCubeMycelium.func_177551_a(facing) : quads) : modelCubeMycelium.func_177551_a(facing);
             }
             else
             {
@@ -55,12 +55,12 @@
 
                     if (snowy)
                     {
-                        if (getBlockAt(blockPos, 0, facing, blockAccess) == Blocks.snow_layer)
+                        if (getBlockAt(blockPos, facing, blockAccess) == Blocks.snow_layer)
                         {
                             return modelCubeGrassSnowy.func_177551_a(facing);
                         }
                     }
-                    else if (getBlockAt(blockPos, -1, facing, blockAccess) == Blocks.grass)
+                    else if (getBlockAt(blockPos.offsetDown(), facing, blockAccess) == Blocks.grass)
                     {
                         return modelCubeGrass.func_177551_a(facing);
                     }
@@ -75,76 +75,10 @@
         }
     }
 
-    private static Block getBlockAt(BlockPos blockPos, int dy, EnumFacing facing, IBlockAccess blockAccess)
+    private static Block getBlockAt(BlockPos blockPos, EnumFacing facing, IBlockAccess blockAccess)
     {
-        int x = blockPos.getX();
-        int y = blockPos.getY() + dy;
-        int z = blockPos.getZ();
-
-        switch (BetterGrass.NamelessClass1391712693.$SwitchMap$net$minecraft$util$EnumFacing[facing.ordinal()])
-        {
-            case 1:
-                --z;
-                break;
-
-            case 2:
-                ++z;
-                break;
-
-            case 3:
-                --x;
-                break;
-
-            case 4:
-                ++x;
-        }
-
-        BlockPos pos = new BlockPos(x, y, z);
+        BlockPos pos = blockPos.offset(facing);
         Block block = blockAccess.getBlockState(pos).getBlock();
         return block;
     }
-
-    static class NamelessClass1391712693
-    {
-        static final int[] $SwitchMap$net$minecraft$util$EnumFacing = new int[EnumFacing.values().length];
-
-        static
-        {
-            try
-            {
-                $SwitchMap$net$minecraft$util$EnumFacing[EnumFacing.NORTH.ordinal()] = 1;
-            }
-            catch (NoSuchFieldError var4)
-            {
-                ;
-            }
-
-            try
-            {
-                $SwitchMap$net$minecraft$util$EnumFacing[EnumFacing.SOUTH.ordinal()] = 2;
-            }
-            catch (NoSuchFieldError var3)
-            {
-                ;
-            }
-
-            try
-            {
-                $SwitchMap$net$minecraft$util$EnumFacing[EnumFacing.WEST.ordinal()] = 3;
-            }
-            catch (NoSuchFieldError var2)
-            {
-                ;
-            }
-
-            try
-            {
-                $SwitchMap$net$minecraft$util$EnumFacing[EnumFacing.EAST.ordinal()] = 4;
-            }
-            catch (NoSuchFieldError var1)
-            {
-                ;
-            }
-        }
-    }
 }
