--- a/net/minecraft/src/BetterSnow.java
+++ b/net/minecraft/src/BetterSnow.java
@@ -51,10 +51,7 @@
     private static boolean hasSnowNeighbours(IBlockAccess blockAccess, BlockPos pos)
     {
         Block blockSnow = Blocks.snow_layer;
-        int x = pos.getX();
-        int y = pos.getY();
-        int z = pos.getZ();
-        return blockAccess.getBlockState(new BlockPos(x - 1, y, z)).getBlock() != blockSnow && blockAccess.getBlockState(new BlockPos(x + 1, y, z)).getBlock() != blockSnow && blockAccess.getBlockState(new BlockPos(x, y, z - 1)).getBlock() != blockSnow && blockAccess.getBlockState(new BlockPos(x, y, z + 1)).getBlock() != blockSnow ? false : blockAccess.getBlockState(new BlockPos(x, y - 1, z)).getBlock().isOpaqueCube();
+        return blockAccess.getBlockState(pos.offsetNorth()).getBlock() != blockSnow && blockAccess.getBlockState(pos.offsetSouth()).getBlock() != blockSnow && blockAccess.getBlockState(pos.offsetWest()).getBlock() != blockSnow && blockAccess.getBlockState(pos.offsetEast()).getBlock() != blockSnow ? false : blockAccess.getBlockState(pos.offsetDown()).getBlock().isOpaqueCube();
     }
 
     private static boolean checkBlock(Block block, IBlockState blockState)
