--- a/net/minecraft/src/BetterGrass.java
+++ b/net/minecraft/src/BetterGrass.java
@@ -81,7 +81,7 @@
         int y = blockPos.getY() + dy;
         int z = blockPos.getZ();
 
-        switch (BetterGrass.NamelessClass1936830976.$SwitchMap$net$minecraft$util$EnumFacing[facing.ordinal()])
+        switch (BetterGrass.NamelessClass1778439638.$SwitchMap$net$minecraft$util$EnumFacing[facing.ordinal()])
         {
             case 1:
                 --z;
@@ -104,7 +104,7 @@
         return block;
     }
 
-    static class NamelessClass1936830976
+    static class NamelessClass1778439638
     {
         static final int[] $SwitchMap$net$minecraft$util$EnumFacing = new int[EnumFacing.values().length];
 
