--- a/net/minecraft/src/BlockPosM.java
+++ b/net/minecraft/src/BlockPosM.java
@@ -157,7 +157,7 @@
                     {
                         if (this.theBlockPosM == null)
                         {
-                            this.theBlockPosM = new BlockPosM(posFromxx.getX(), posFrom.getY(), posFromx.getZ(), 3);
+                            this.theBlockPosM = new BlockPosM(posFrom.getX(), posFrom.getY(), posFrom.getZ(), 3);
                             return this.theBlockPosM;
                         }
                         else if (this.theBlockPosM.equals(posTo))
