--- a/net/minecraft/src/BlockPosM.java
+++ b/net/minecraft/src/BlockPosM.java
@@ -113,7 +113,7 @@
      */
     public BlockPos offset(EnumFacing facing, int n)
     {
-        return n == 1 ? this.offset(facing) : super.offset(facing, 1);
+        return n == 1 ? this.offset(facing) : super.offset(facing, n);
     }
 
     private void update()
@@ -150,7 +150,7 @@
                     {
                         if (this.theBlockPosM == null)
                         {
-                            this.theBlockPosM = new BlockPosM(posFromxx.getX(), posFrom.getY(), posFromx.getZ(), 3);
+                            this.theBlockPosM = new BlockPosM(posFrom.getX(), posFrom.getY(), posFrom.getZ(), 3);
                             return this.theBlockPosM;
                         }
                         else if (this.theBlockPosM.equals(posTo))
