--- a/net/minecraft/src/Config.java
+++ b/net/minecraft/src/Config.java
@@ -1385,6 +1385,12 @@
 
     public static void checkDisplaySettings()
     {
+        /** MINECRIFT */
+        // Dont allow changes to the display in Direct Mode
+        if (Minecraft.getMinecraft().isDirectMode)
+            return;
+        /** END MINECRIFT */
+
         if (getAntialiasingLevel() > 0)
         {
             int samples = getAntialiasingLevel();
@@ -1461,6 +1467,12 @@
 
     public static void checkDisplayMode()
     {
+        /** MINECRIFT */
+
+        // Dont allow changes to the display in Direct Mode
+        if (Minecraft.getMinecraft().isDirectMode)
+            return;
+
         try
         {
             if (minecraft.isFullScreen())
@@ -1493,31 +1505,35 @@
                 }
 
                 Display.setDisplayMode(newMode);
-                minecraft.displayWidth = Display.getDisplayMode().getWidth();
-                minecraft.displayHeight = Display.getDisplayMode().getHeight();
-
-                if (minecraft.displayWidth <= 0)
-                {
-                    minecraft.displayWidth = 1;
-                }
+                Minecraft.getMinecraft().reinitFramebuffers = true;
 
-                if (minecraft.displayHeight <= 0)
-                {
-                    minecraft.displayHeight = 1;
-                }
-
-                if (minecraft.currentScreen != null)
-                {
-                    ScaledResolution sr = new ScaledResolution(minecraft, minecraft.displayWidth, minecraft.displayHeight);
-                    int sw = sr.getScaledWidth();
-                    int sh = sr.getScaledHeight();
-                    minecraft.currentScreen.setWorldAndResolution(minecraft, sw, sh);
-                }
+//                minecraft.displayWidth = Display.getDisplayMode().getWidth();
+//                minecraft.displayHeight = Display.getDisplayMode().getHeight();
+//
+//                if (minecraft.displayWidth <= 0)
+//                {
+//                    minecraft.displayWidth = 1;
+//                }
+//
+//                if (minecraft.displayHeight <= 0)
+//                {
+//                    minecraft.displayHeight = 1;
+//                }
+//
+//                if (minecraft.currentScreen != null)
+//                {
+//                    ScaledResolution sr = new ScaledResolution(minecraft, minecraft.displayWidth, minecraft.displayHeight);
+//                    int sw = sr.getScaledWidth();
+//                    int sh = sr.getScaledHeight();
+//                    minecraft.currentScreen.setWorldAndResolution(minecraft, sw, sh);
+//                }
+//
+//                minecraft.loadingScreen = new LoadingScreenRenderer(minecraft);
+//                updateFramebufferSize();
 
-                minecraft.loadingScreen = new LoadingScreenRenderer(minecraft);
-                updateFramebufferSize();
                 Display.setFullscreen(true);
                 minecraft.gameSettings.updateVSync();
+                Display.update();
                 GL11.glEnable(GL11.GL_TEXTURE_2D);
             }
             else
@@ -1538,6 +1554,7 @@
         {
             var6.printStackTrace();
         }
+        /** END MINECRIFT */
     }
 
     private static void updateFramebufferSize()
