--- a/net/minecraft/src/ConnectedProperties.java
+++ b/net/minecraft/src/ConnectedProperties.java
@@ -1087,45 +1087,46 @@
             for (int icons = 0; icons < tileNames.length; ++icons)
             {
                 String iconName = tileNames[icons];
-                String fullName = iconName;
+                ResourceLocation resLoc = new ResourceLocation(iconName);
+                String domain = resLoc.getResourceDomain();
+                String path = resLoc.getResourcePath();
 
-                if (!iconName.contains("/"))
+                if (!path.contains("/"))
                 {
-                    fullName = "textures/blocks/" + iconName;
+                    path = "textures/blocks/" + path;
                 }
 
-                String fileName = fullName + ".png";
-                ResourceLocation loc = new ResourceLocation(fileName);
-                boolean exists = Config.hasResource(loc);
+                String filePath = path + ".png";
+                ResourceLocation locFile = new ResourceLocation(domain, filePath);
+                boolean exists = Config.hasResource(locFile);
 
                 if (!exists)
                 {
-                    Config.warn("File not found: " + fileName);
+                    Config.warn("File not found: " + filePath);
                 }
 
-                String shortName = fullName;
+                String prefixTextures = "textures/";
+                String pathSprite = path;
 
-                if (fullName.startsWith("textures/"))
+                if (path.startsWith(prefixTextures))
                 {
-                    shortName = fullName.substring("textures/".length());
+                    pathSprite = path.substring(prefixTextures.length());
                 }
 
-                ResourceLocation locSprite = new ResourceLocation(shortName);
+                ResourceLocation locSprite = new ResourceLocation(domain, pathSprite);
                 TextureAtlasSprite icon = textureMap.func_174942_a(locSprite);
                 iconList.add(icon);
             }
 
-            TextureAtlasSprite[] var12 = (TextureAtlasSprite[])((TextureAtlasSprite[])iconList.toArray(new TextureAtlasSprite[iconList.size()]));
-            return var12;
+            TextureAtlasSprite[] var15 = (TextureAtlasSprite[])((TextureAtlasSprite[])iconList.toArray(new TextureAtlasSprite[iconList.size()]));
+            return var15;
         }
     }
 
-    public boolean matchesBlock(Block block)
+    public boolean matchesBlock(int blockId)
     {
         if (this.matchBlocks != null && this.matchBlocks.length > 0)
         {
-            int blockId = Block.getIdFromBlock(block);
-
             for (int i = 0; i < this.matchBlocks.length; ++i)
             {
                 int matchId = this.matchBlocks[i];
