--- a/net/minecraft/src/ConnectedTextures.java
+++ b/net/minecraft/src/ConnectedTextures.java
@@ -264,7 +264,7 @@
         }
         else
         {
-            switch (ConnectedTextures.NamelessClass746758842.$SwitchMap$net$minecraft$util$EnumFacing[facing.ordinal()])
+            switch (ConnectedTextures.NamelessClass1360927152.$SwitchMap$net$minecraft$util$EnumFacing[facing.ordinal()])
             {
                 case 1:
                     return 0;
@@ -925,7 +925,11 @@
     {
         IBlockState neighbourState = iblockaccess.getBlockState(blockPos);
 
-        if (cp.connect == 2)
+        if (blockState == neighbourState)
+        {
+            return true;
+        }
+        else if (cp.connect == 2)
         {
             if (neighbourState == null)
             {
@@ -937,18 +941,19 @@
             }
             else
             {
-                TextureAtlasSprite neighbourIcon = getNeighbourIcon(neighbourState, side);
+                TextureAtlasSprite neighbourIcon = getNeighbourIcon(iblockaccess, blockPos, neighbourState, side);
                 return neighbourIcon == icon;
             }
         }
         else
         {
-            return cp.connect == 3 ? (neighbourState == null ? false : (neighbourState == AIR_DEFAULT_STATE ? false : neighbourState.getBlock().getMaterial() == blockState.getBlock().getMaterial())) : neighbourState == blockState;
+            return cp.connect == 3 ? (neighbourState == null ? false : (neighbourState == AIR_DEFAULT_STATE ? false : neighbourState.getBlock().getMaterial() == blockState.getBlock().getMaterial())) : false;
         }
     }
 
-    private static TextureAtlasSprite getNeighbourIcon(IBlockState neighbourState, int side)
+    private static TextureAtlasSprite getNeighbourIcon(IBlockAccess iblockaccess, BlockPos blockPos, IBlockState neighbourState, int side)
     {
+        neighbourState = neighbourState.getBlock().getActualState(neighbourState, iblockaccess, blockPos);
         IBakedModel model = Minecraft.getMinecraft().getBlockRendererDispatcher().func_175023_a().func_178125_b(neighbourState);
 
         if (model == null)
@@ -962,8 +967,8 @@
 
             if (quads.size() > 0)
             {
-                BakedQuad var8 = (BakedQuad)quads.get(0);
-                return var8.getSprite();
+                BakedQuad var10 = (BakedQuad)quads.get(0);
+                return var10.getSprite();
             }
             else
             {
@@ -1663,7 +1668,7 @@
         }
     }
 
-    static class NamelessClass746758842
+    static class NamelessClass1360927152
     {
         static final int[] $SwitchMap$net$minecraft$util$EnumFacing = new int[EnumFacing.values().length];
 
