--- a/net/minecraft/src/CustomSky.java
+++ b/net/minecraft/src/CustomSky.java
@@ -9,7 +9,6 @@
 import net.minecraft.client.renderer.texture.TextureManager;
 import net.minecraft.util.ResourceLocation;
 import net.minecraft.world.World;
-import org.lwjgl.opengl.GL11;
 
 public class CustomSky
 {
@@ -155,7 +154,7 @@
                             }
                         }
 
-                        clearBlend(rainBrightness);
+                        Blender.clearBlend(rainBrightness);
                     }
                 }
             }
@@ -187,12 +186,4 @@
             }
         }
     }
-
-    private static void clearBlend(float rainBrightness)
-    {
-        GL11.glDisable(GL11.GL_ALPHA_TEST);
-        GL11.glEnable(GL11.GL_BLEND);
-        GL11.glBlendFunc(GL11.GL_SRC_ALPHA, GL11.GL_ONE);
-        GL11.glColor4f(1.0F, 1.0F, 1.0F, rainBrightness);
-    }
 }
