--- a/net/minecraft/src/NaturalTextures.java
+++ b/net/minecraft/src/NaturalTextures.java
@@ -1,7 +1,5 @@
 package net.minecraft.src;
 
-import java.io.FileNotFoundException;
-import java.io.InputStream;
 import java.util.ArrayList;
 import java.util.List;
 import net.minecraft.client.renderer.texture.TextureAtlasSprite;
@@ -15,93 +13,6 @@
     public static void update()
     {
         propertiesByIndex = new NaturalProperties[0];
-
-        if (Config.isNaturalTextures())
-        {
-            String fileName = "optifine/natural.properties";
-
-            try
-            {
-                ResourceLocation e = new ResourceLocation(fileName);
-
-                if (!Config.hasResource(e))
-                {
-                    Config.dbg("NaturalTextures: configuration \"" + fileName + "\" not found");
-                    propertiesByIndex = makeDefaultProperties();
-                    return;
-                }
-
-                InputStream in = Config.getResourceStream(e);
-                ArrayList list = new ArrayList(256);
-                String configStr = Config.readInputStream(in);
-                in.close();
-                String[] configLines = Config.tokenize(configStr, "\n\r");
-                Config.dbg("Natural Textures: Parsing configuration \"" + fileName + "\"");
-                TextureMap textureMapBlocks = TextureUtils.getTextureMapBlocks();
-
-                for (int i = 0; i < configLines.length; ++i)
-                {
-                    String line = configLines[i].trim();
-
-                    if (!line.startsWith("#"))
-                    {
-                        String[] strs = Config.tokenize(line, "=");
-
-                        if (strs.length != 2)
-                        {
-                            Config.warn("Natural Textures: Invalid \"" + fileName + "\" line: " + line);
-                        }
-                        else
-                        {
-                            String key = strs[0].trim();
-                            String type = strs[1].trim();
-                            TextureAtlasSprite ts = textureMapBlocks.getSpriteSafe("minecraft:blocks/" + key);
-
-                            if (ts == null)
-                            {
-                                Config.warn("Natural Textures: Texture not found: \"" + fileName + "\" line: " + line);
-                            }
-                            else
-                            {
-                                int tileNum = ts.getIndexInMap();
-
-                                if (tileNum < 0)
-                                {
-                                    Config.warn("Natural Textures: Invalid \"" + fileName + "\" line: " + line);
-                                }
-                                else
-                                {
-                                    NaturalProperties props = new NaturalProperties(type);
-
-                                    if (props.isValid())
-                                    {
-                                        while (list.size() <= tileNum)
-                                        {
-                                            list.add((Object)null);
-                                        }
-
-                                        list.set(tileNum, props);
-                                        Config.dbg("NaturalTextures: " + key + " = " + type);
-                                    }
-                                }
-                            }
-                        }
-                    }
-                }
-
-                propertiesByIndex = (NaturalProperties[])((NaturalProperties[])list.toArray(new NaturalProperties[list.size()]));
-            }
-            catch (FileNotFoundException var16)
-            {
-                Config.warn("NaturalTextures: configuration \"" + fileName + "\" not found");
-                propertiesByIndex = makeDefaultProperties();
-                return;
-            }
-            catch (Exception var17)
-            {
-                var17.printStackTrace();
-            }
-        }
     }
 
     public static NaturalProperties getNaturalProperties(TextureAtlasSprite icon)
