--- a/net/minecraft/src/RandomMobs.java
+++ b/net/minecraft/src/RandomMobs.java
@@ -80,7 +80,7 @@
         }
         else
         {
-            ResourceLocation var5;
+            ResourceLocation props;
 
             try
             {
@@ -108,28 +108,29 @@
                 EntityLiving entityLiving = (EntityLiving)entity;
                 String name = loc.getResourcePath();
 
-                if (!name.startsWith("textures/entity/"))
+                if (name.startsWith("textures/entity/"))
                 {
-                    ResourceLocation props1 = loc;
-                    return props1;
-                }
+                    RandomMobsProperties props1 = getProperties(loc);
+                    ResourceLocation var5;
 
-                RandomMobsProperties props = getProperties(loc);
+                    if (props1 == null)
+                    {
+                        var5 = loc;
+                        return var5;
+                    }
 
-                if (props != null)
-                {
-                    var5 = props.getTextureLocation(loc, entityLiving);
+                    var5 = props1.getTextureLocation(loc, entityLiving);
                     return var5;
                 }
 
-                var5 = loc;
+                props = loc;
             }
             finally
             {
                 working = false;
             }
 
-            return var5;
+            return props;
         }
     }
 
