--- a/net/minecraft/src/RandomMobs.java
+++ b/net/minecraft/src/RandomMobs.java
@@ -105,7 +105,7 @@
         }
         else
         {
-            ResourceLocation uuidLow;
+            ResourceLocation name;
 
             try
             {
@@ -123,38 +123,37 @@
                 }
 
                 Entity entity = renderGlobal.renderedEntity;
-                ResourceLocation name1;
 
                 if (entity == null)
                 {
-                    name1 = loc;
-                    return name1;
+                    name = loc;
+                    return name;
                 }
 
-                if (!(entity instanceof EntityLiving))
+                if (entity instanceof EntityLiving)
                 {
-                    name1 = loc;
-                    return name1;
-                }
+                    String name1 = loc.getResourcePath();
 
-                String name = loc.getResourcePath();
+                    if (!name1.startsWith("textures/entity/"))
+                    {
+                        ResourceLocation uuidLow1 = loc;
+                        return uuidLow1;
+                    }
 
-                if (name.startsWith("textures/entity/"))
-                {
-                    long uuidLow1 = entity.getUniqueID().getLeastSignificantBits();
-                    int id = (int)(uuidLow1 & 2147483647L);
+                    long uuidLow = entity.getUniqueID().getLeastSignificantBits();
+                    int id = (int)(uuidLow & 2147483647L);
                     ResourceLocation var6 = getTextureLocation(loc, id);
                     return var6;
                 }
 
-                uuidLow = loc;
+                name = loc;
             }
             finally
             {
                 working = false;
             }
 
-            return uuidLow;
+            return name;
         }
     }
 
