--- a/net/minecraft/src/TextureAnimation.java
+++ b/net/minecraft/src/TextureAnimation.java
@@ -3,6 +3,7 @@
 import java.nio.ByteBuffer;
 import java.util.Properties;
 import net.minecraft.client.renderer.GLAllocation;
+import net.minecraft.client.renderer.GlStateManager;
 import org.lwjgl.opengl.GL11;
 
 public class TextureAnimation
@@ -138,7 +139,7 @@
             else
             {
                 this.imageData.position(offset);
-                GL11.glBindTexture(GL11.GL_TEXTURE_2D, this.dstTextId);
+                GlStateManager.func_179144_i(this.dstTextId);
                 GL11.glTexSubImage2D(GL11.GL_TEXTURE_2D, 0, this.dstX, this.dstY, this.frameWidth, this.frameHeight, GL11.GL_RGBA, GL11.GL_UNSIGNED_BYTE, this.imageData);
                 return true;
             }
