--- a/net/minecraft/src/WrUpdateThread.java
+++ b/net/minecraft/src/WrUpdateThread.java
@@ -46,7 +46,7 @@
             var8.printStackTrace();
         }
 
-        WrUpdateThread.ThreadUpdateListener updateListener = new WrUpdateThread.ThreadUpdateListener((WrUpdateThread.NamelessClass252209705)null);
+        WrUpdateThread.ThreadUpdateListener updateListener = new WrUpdateThread.ThreadUpdateListener((WrUpdateThread.NamelessClass74792162)null);
 
         while (!Thread.interrupted() && !this.terminated)
         {
@@ -367,7 +367,7 @@
         this.terminated = true;
     }
 
-    static class NamelessClass252209705
+    static class NamelessClass74792162
     {
     }
 
@@ -407,7 +407,7 @@
             this.updateControl = updateControl;
         }
 
-        ThreadUpdateControl(WrUpdateThread.NamelessClass252209705 x1)
+        ThreadUpdateControl(WrUpdateThread.NamelessClass74792162 x1)
         {
             this();
         }
@@ -419,7 +419,7 @@
 
         private ThreadUpdateListener()
         {
-            this.tuc = WrUpdateThread.this.new ThreadUpdateControl((WrUpdateThread.NamelessClass252209705)null);
+            this.tuc = WrUpdateThread.this.new ThreadUpdateControl((WrUpdateThread.NamelessClass74792162)null);
         }
 
         public void updating(IWrUpdateControl uc)
@@ -428,7 +428,7 @@
             WrUpdateThread.this.checkCanWork(this.tuc);
         }
 
-        ThreadUpdateListener(WrUpdateThread.NamelessClass252209705 x1)
+        ThreadUpdateListener(WrUpdateThread.NamelessClass74792162 x1)
         {
             this();
         }
