--- a/net/minecraft/src/xp.java
+++ b/net/minecraft/src/xp.java
@@ -4,11 +4,11 @@
 import net.minecraft.entity.Entity;
 import net.minecraft.entity.EntityLivingBase;
 
-class xp implements Predicate
+public class xp implements Predicate
 {
-    final EntityLivingBase a;
+    public final EntityLivingBase a;
 
-    xp(EntityLivingBase p_i46455_1_)
+    public xp(EntityLivingBase p_i46455_1_)
     {
         this.a = p_i46455_1_;
     }
