--- a/net/minecraft/entity/passive/EntityRabbit.java
+++ b/net/minecraft/entity/passive/EntityRabbit.java
@@ -498,9 +498,9 @@
         private EntityRabbit field_179511_d = EntityRabbit.this;
         private static final String __OBFID = "CL_00002238";
 
-        public AIAvoidEntity(Class p_i46386_2_, float p_i46386_3_, double p_i46386_4_, double p_i46386_6_)
+        public AIAvoidEntity(Class p_i46437_2_, float p_i46437_3_, double p_i46437_4_, double p_i46437_6_)
         {
-            super(EntityRabbit.this, p_i46386_2_, p_i46386_3_, p_i46386_4_, p_i46386_6_);
+            super(EntityRabbit.this, p_i46437_2_, p_i46437_3_, p_i46437_4_, p_i46437_6_);
         }
 
         public void updateTask()
