--- a/net/minecraft/client/main/GameConfiguration.java
+++ b/net/minecraft/client/main/GameConfiguration.java
@@ -91,12 +91,12 @@
         public final Proxy field_178751_c;
         private static final String __OBFID = "CL_00001913";
 
-        public UserInformation(Session p_i46376_1_, PropertyMap p_i46376_2_, PropertyMap p_i46376_3_, Proxy p_i46376_4_)
+        public UserInformation(Session p_i46439_1_, PropertyMap p_i46439_2_, PropertyMap p_i46439_3_, Proxy p_i46439_4_)
         {
-            this.field_178752_a = p_i46376_1_;
-            this.field_178750_b = p_i46376_2_;
-            this.c = p_i46376_3_;
-            this.field_178751_c = p_i46376_4_;
+            this.field_178752_a = p_i46439_1_;
+            this.field_178750_b = p_i46439_2_;
+            this.c = p_i46439_3_;
+            this.field_178751_c = p_i46439_4_;
         }
     }
 }
