--- a/net/minecraft/client/main/Main.java
+++ b/net/minecraft/client/main/Main.java
@@ -59,7 +59,7 @@
 
     public static void main(String[] p_main_0_)
     {
-        System.setProperty("java.net.preferIPv4Stack", "true");
+        System.setProperty("java.net.preferIPv4Stack", "true");	
         OptionParser var1 = new OptionParser();
         var1.allowsUnrecognizedOptions();
         var1.accepts("demo");
