--- a/net/minecraft/client/renderer/chunk/ChunkCompileTaskGenerator.java
+++ b/net/minecraft/client/renderer/chunk/ChunkCompileTaskGenerator.java
@@ -17,7 +17,6 @@
     private ChunkCompileTaskGenerator.Status field_178548_g;
     private boolean field_178554_h;
     private static final String __OBFID = "CL_00002466";
-    private static final String __OBFID = "CL_00002466";
 
     public ChunkCompileTaskGenerator(RenderChunk p_i46208_1_, ChunkCompileTaskGenerator.Type p_i46208_2_)
     {
@@ -141,7 +140,6 @@
         private static final String __OBFID = "CL_00002465";
 
         private static final ChunkCompileTaskGenerator.Status[] $VALUES$ = new ChunkCompileTaskGenerator.Status[]{PENDING, COMPILING, UPLOADING, DONE};
-        private static final String __OBFID = "CL_00002465";
 
         private Status(String p_i46428_1_, int p_i46428_2_, String p_i46207_1_, int p_i46207_2_) {}
     }
@@ -154,7 +152,6 @@
         private static final String __OBFID = "CL_00002464";
 
         private static final ChunkCompileTaskGenerator.Type[] $VALUES$ = new ChunkCompileTaskGenerator.Type[]{REBUILD_CHUNK, RESORT_TRANSPARENCY};
-        private static final String __OBFID = "CL_00002464";
 
         private Type(String p_i46396_1_, int p_i46396_2_, String p_i46206_1_, int p_i46206_2_) {}
     }
