--- a/net/minecraft/client/renderer/chunk/RenderChunk.java
+++ b/net/minecraft/client/renderer/chunk/RenderChunk.java
@@ -47,7 +47,6 @@
     private boolean field_178593_n;
     private static final String __OBFID = "CL_00002452";
     private BlockPos[] positionOffsets16;
-    private static final String __OBFID = "CL_00002452";
 
     public RenderChunk(World worldIn, RenderGlobal renderGlobalIn, BlockPos blockPosIn, int indexIn)
     {
