--- a/net/minecraft/src/BlockUtils.java
+++ b/net/minecraft/src/BlockUtils.java
@@ -12,19 +12,11 @@
     {
         if (directAccessValid)
         {
-            try
-            {
-                block.setLightOpacity(opacity);
-                return;
-            }
-            catch (IllegalAccessError var3)
-            {
-                directAccessValid = false;
+            directAccessValid = false;
 
-                if (!ForgeBlock_setLightOpacity.exists())
-                {
-                    throw var3;
-                }
+            if (!ForgeBlock_setLightOpacity.exists())
+            {
+                throw new IllegalAccessError("Block.class - setLightOpacity field doesn't exist!");
             }
         }
 
