--- a/net/minecraft/src/EntityUtils.java
+++ b/net/minecraft/src/EntityUtils.java
@@ -16,18 +16,11 @@
     {
         if (directEntityAge)
         {
-            try
-            {
-                return elb.entityAge;
-            }
-            catch (IllegalAccessError var2)
-            {
-                directEntityAge = false;
+            directEntityAge = false;
 
-                if (!ForgeEntityLivingBase_entityAge.exists())
-                {
-                    throw var2;
-                }
+            if (!ForgeEntityLivingBase_entityAge.exists())
+            {
+                throw new IllegalAccessError("EntityLivingBase.class - entityAge field doesn't exist!");
             }
         }
 
@@ -38,19 +31,11 @@
     {
         if (directEntityAge)
         {
-            try
-            {
-                elb.entityAge = age;
-                return;
-            }
-            catch (IllegalAccessError var3)
-            {
-                directEntityAge = false;
+            directEntityAge = false;
 
-                if (!ForgeEntityLivingBase_entityAge.exists())
-                {
-                    throw var3;
-                }
+            if (!ForgeEntityLivingBase_entityAge.exists())
+            {
+                throw new IllegalAccessError("EntityLivingBase.class - entityAge field doesn't exist!");
             }
         }
 
@@ -61,19 +46,11 @@
     {
         if (directDespawnEntity)
         {
-            try
-            {
-                el.despawnEntity();
-                return;
-            }
-            catch (IllegalAccessError var2)
-            {
-                directDespawnEntity = false;
+            directDespawnEntity = false;
 
-                if (!ForgeEntityLiving_despawnEntity.exists())
-                {
-                    throw var2;
-                }
+            if (!ForgeEntityLiving_despawnEntity.exists())
+            {
+                throw new IllegalAccessError("EntityLiving.class - despawnEntity method doesn't exist!");
             }
         }
 
