--- a/net/minecraft/src/ResourceUtils.java
+++ b/net/minecraft/src/ResourceUtils.java
@@ -13,18 +13,11 @@
     {
         if (directAccessValid)
         {
-            try
-            {
-                return arp.resourcePackFile;
-            }
-            catch (IllegalAccessError var2)
-            {
-                directAccessValid = false;
+            directAccessValid = false;
 
-                if (!ForgeAbstractResourcePack_resourcePackFile.exists())
-                {
-                    throw var2;
-                }
+            if (!ForgeAbstractResourcePack_resourcePackFile.exists())
+            {
+                throw new IllegalAccessError("AbstractResourcePack.class - resourcePackFile field doesn't exist!");
             }
         }
 
