--- a/net/minecraft/src/WorldRendererThreaded.java
+++ b/net/minecraft/src/WorldRendererThreaded.java
@@ -16,6 +16,7 @@
 import net.minecraft.tileentity.TileEntity;
 import net.minecraft.util.AxisAlignedBB;
 import net.minecraft.util.MathHelper;
+import net.minecraft.util.Vec3;
 import net.minecraft.world.ChunkCache;
 import net.minecraft.world.World;
 import net.minecraft.world.chunk.Chunk;
@@ -167,7 +168,10 @@
                                     {
                                         hasRenderedBlocks |= hashset1.renderBlockByRenderType(block, x, y, z);
 
-                                        if (block.getRenderType() == 0 && x == viewEntityPosX && y == viewEntityPosY && z == viewEntityPosZ)
+                                        //Vive
+                                        Vec3 cam =  Minecraft.getMinecraft().entityRenderer.getCameraBlock();
+                                        if (block.getRenderType() == 0 && x ==cam.xCoord && y == cam.yCoord && z == cam.zCoord)
+                                       	//END Vive
                                         {
                                             hashset1.setRenderFromInside(true);
                                             hashset1.setRenderAllFaces(true);
