--- a/net/minecraft/crash/CrashReport.java
+++ b/net/minecraft/crash/CrashReport.java
@@ -43,7 +43,6 @@
     private StackTraceElement[] stacktrace = new StackTraceElement[0];
     private static final String __OBFID = "CL_00000990";
     private boolean reported = false;
-    private static final String __OBFID = "CL_00000990";
 
     public CrashReport(String par1Str, Throwable par2Throwable)
     {
@@ -61,59 +60,38 @@
         this.theReportCategory.addCrashSectionCallable("Minecraft Version", new Callable()
         {
             private static final String __OBFID = "CL_00001197";
-            private static final String __OBFID = "CL_00001197";
             public String call()
             {
                 return "1.7.10";
             }
-            public Object call() throws Exception
-            {
-                return this.call();
-            }
         });
         this.theReportCategory.addCrashSectionCallable("Operating System", new Callable()
         {
             private static final String __OBFID = "CL_00001222";
-            private static final String __OBFID = "CL_00001222";
             public String call()
             {
                 return System.getProperty("os.name") + " (" + System.getProperty("os.arch") + ") version " + System.getProperty("os.version");
             }
-            public Object call() throws Exception
-            {
-                return this.call();
-            }
         });
         this.theReportCategory.addCrashSectionCallable("Java Version", new Callable()
         {
             private static final String __OBFID = "CL_00001248";
-            private static final String __OBFID = "CL_00001248";
             public String call()
             {
                 return System.getProperty("java.version") + ", " + System.getProperty("java.vendor");
             }
-            public Object call() throws Exception
-            {
-                return this.call();
-            }
         });
         this.theReportCategory.addCrashSectionCallable("Java VM Version", new Callable()
         {
             private static final String __OBFID = "CL_00001275";
-            private static final String __OBFID = "CL_00001275";
             public String call()
             {
                 return System.getProperty("java.vm.name") + " (" + System.getProperty("java.vm.info") + "), " + System.getProperty("java.vm.vendor");
             }
-            public Object call() throws Exception
-            {
-                return this.call();
-            }
         });
         this.theReportCategory.addCrashSectionCallable("Memory", new Callable()
         {
             private static final String __OBFID = "CL_00001302";
-            private static final String __OBFID = "CL_00001302";
             public String call()
             {
                 Runtime var1 = Runtime.getRuntime();
@@ -125,15 +103,10 @@
                 long var12 = var6 / 1024L / 1024L;
                 return var6 + " bytes (" + var12 + " MB) / " + var4 + " bytes (" + var10 + " MB) up to " + var2 + " bytes (" + var8 + " MB)";
             }
-            public Object call() throws Exception
-            {
-                return this.call();
-            }
         });
         this.theReportCategory.addCrashSectionCallable("JVM Flags", new Callable()
         {
             private static final String __OBFID = "CL_00001329";
-            private static final String __OBFID = "CL_00001329";
             public String call()
             {
                 RuntimeMXBean var1 = ManagementFactory.getRuntimeMXBean();
@@ -159,15 +132,10 @@
 
                 return String.format("%d total; %s", new Object[] {Integer.valueOf(var3), var4.toString()});
             }
-            public Object call() throws Exception
-            {
-                return this.call();
-            }
         });
         this.theReportCategory.addCrashSectionCallable("AABB Pool Size", new Callable()
         {
             private static final String __OBFID = "CL_00001355";
-            private static final String __OBFID = "CL_00001355";
             public String call()
             {
                 byte var1 = 0;
@@ -178,23 +146,14 @@
                 int var6 = var5 / 1024 / 1024;
                 return var1 + " (" + var2 + " bytes; " + var3 + " MB) allocated, " + var4 + " (" + var5 + " bytes; " + var6 + " MB) used";
             }
-            public Object call() throws Exception
-            {
-                return this.call();
-            }
         });
         this.theReportCategory.addCrashSectionCallable("IntCache", new Callable()
         {
             private static final String __OBFID = "CL_00001382";
-            private static final String __OBFID = "CL_00001382";
             public String call() throws SecurityException, NoSuchFieldException, IllegalAccessException, IllegalArgumentException
             {
                 return IntCache.getCacheSizes();
             }
-            public Object call() throws Exception
-            {
-                return this.call();
-            }
         });
 
         if (Reflector.FMLCommonHandler_enhanceCrashReport.exists())
