--- a/net/minecraft/client/resources/ResourcePackRepository.java
+++ b/net/minecraft/client/resources/ResourcePackRepository.java
@@ -200,7 +200,6 @@
 
         final File var4 = new File(this.field_148534_e, var3);
         this.field_177321_h.lock();
-        ListenableFuture var9;
 
         try
         {
@@ -254,14 +253,13 @@
                     var8.setException(p_onFailure_1_);
                 }
             });
-            var9 = this.field_177322_i;
+            ListenableFuture var9 = this.field_177322_i;
+            return var9;
         }
         finally
         {
             this.field_177321_h.unlock();
         }
-
-        return var9;
     }
 
     public ListenableFuture func_177319_a(File p_177319_1_)
