--- a/net/minecraft/entity/EntityLiving.java
+++ b/net/minecraft/entity/EntityLiving.java
@@ -84,7 +84,6 @@
     private Entity leashedToEntity;
     private NBTTagCompound leashNBTTag;
     private static final String __OBFID = "CL_00001550";
-    private static final String __OBFID = "CL_00001550";
 
     public EntityLiving(World worldIn)
     {
@@ -1404,7 +1403,6 @@
         private static final String __OBFID = "CL_00002255";
 
         private static final EntityLiving.SpawnPlacementType[] $VALUES$ = new EntityLiving.SpawnPlacementType[]{ON_GROUND, IN_AIR, IN_WATER};
-        private static final String __OBFID = "CL_00002255";
 
         private SpawnPlacementType(String p_i46384_1_, int p_i46384_2_, String p_i45893_1_, int p_i45893_2_) {}
     }
