--- a/net/minecraft/client/network/NetHandlerPlayClient.java
+++ b/net/minecraft/client/network/NetHandlerPlayClient.java
@@ -2,6 +2,7 @@
 
 import com.google.common.base.Charsets;
 import com.mojang.authlib.GameProfile;
+import com.mtbs3d.minecrift.gameplay.VRPlayer;
 import io.netty.buffer.ByteBuf;
 import io.netty.buffer.Unpooled;
 import io.netty.util.concurrent.GenericFutureListener;
@@ -290,6 +291,11 @@
         this.gameController.playerController.setGameType(packetIn.func_149198_e());
         this.gameController.gameSettings.sendSettingsToServer();
         this.netManager.scheduleOutboundPacket(new C17PacketCustomPayload("MC|Brand", ClientBrandRetriever.getClientModName().getBytes(Charsets.UTF_8)), new GenericFutureListener[0]);
+        // VIVE START - ask server if it's running this mod
+        this.gameController.vrPlayer.restrictedViveClient = true;
+        String please = "please";
+        this.netManager.scheduleOutboundPacket(new C17PacketCustomPayload("MC|Vive", please.getBytes(Charsets.UTF_8)), new GenericFutureListener[0]);
+        // VIVE END - ask server if it's running this mod
     }
 
     /**
@@ -739,6 +745,9 @@
     public void handleDisconnect(S40PacketDisconnect packetIn)
     {
         this.netManager.closeChannel(packetIn.func_149165_c());
+        // VIVE START - no longer on a vanilla server, reset restricted state
+        VRPlayer.get().restrictedViveClient = false;
+        // VIVE END - no longer on a vanilla server, reset restricted state
     }
 
     /**
@@ -1663,6 +1672,12 @@
                 }, I18n.format("multiplayer.texturePrompt.line1", new Object[0]), I18n.format("multiplayer.texturePrompt.line2", new Object[0]), 0));
             }
         }
+        // VIVE START - server told us that it has this mod too, allow extended reach, etc.
+        else if ("MC|ViveOK".equals(packetIn.func_149169_c()))
+        {
+            this.gameController.vrPlayer.restrictedViveClient = false;
+        }
+        // VIVE END
     }
 
     /**
