--- a/net/minecraft/client/network/NetHandlerPlayClient.java
+++ b/net/minecraft/client/network/NetHandlerPlayClient.java
@@ -2,6 +2,11 @@
 
 import com.google.common.base.Charsets;
 import com.mojang.authlib.GameProfile;
+import com.mtbs3d.minecrift.api.NetworkHelper;
+import com.mtbs3d.minecrift.api.NetworkHelper.PacketDiscriminators;
+import com.mtbs3d.minecrift.provider.OpenVRPlayer;
+
+import net.minecraft.src.Reflector;
 import io.netty.buffer.ByteBuf;
 import io.netty.buffer.Unpooled;
 import io.netty.util.concurrent.GenericFutureListener;
@@ -193,6 +198,8 @@
 import net.minecraft.world.storage.ISaveHandler;
 import net.minecraft.world.storage.MapData;
 import net.minecraft.world.storage.MapStorage;
+
+import org.apache.commons.lang3.ArrayUtils;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
@@ -290,6 +297,12 @@
         this.gameController.playerController.setGameType(packetIn.func_149198_e());
         this.gameController.gameSettings.sendSettingsToServer();
         this.netManager.scheduleOutboundPacket(new C17PacketCustomPayload("MC|Brand", ClientBrandRetriever.getClientModName().getBytes(Charsets.UTF_8)), new GenericFutureListener[0]);
+        // VIVE START - ask server if it's running this mod
+        this.gameController.vrPlayer.setFreeMoveMode(true);
+        byte[] version = Minecraft.getMinecraft().minecriftVerString.getBytes(Charsets.UTF_8);
+        this.addToSendQueue(new C17PacketCustomPayload("MC|Vive|Version", version));
+        this.addToSendQueue(NetworkHelper.getVivecraftClientPacket(PacketDiscriminators.VERSION, version));
+        // VIVE END - ask server if it's running this mod
     }
 
     /**
@@ -739,6 +752,10 @@
     public void handleDisconnect(S40PacketDisconnect packetIn)
     {
         this.netManager.closeChannel(packetIn.func_149165_c());
+        // VIVE START - no longer on a vanilla server, reset restricted state
+    	this.gameController.vrPlayer.noTeleportClient = true;
+        OpenVRPlayer.get().setFreeMoveMode(true);
+        // VIVE END - no longer on a vanilla server, reset restricted state
     }
 
     /**
@@ -1297,6 +1314,11 @@
                 {
                     var2.readFromNBT(packetIn.getNbtCompound());
                 }
+                /** MINECRIFT FORGE **/
+                else if (Reflector.forgeExists()) {
+                    Reflector.callVoid(var2, Reflector.ForgeTileEntity_onDataPacket, netManager, packetIn);
+                }
+                /** END MINECRIFT FORGE **/
             }
         }
     }
@@ -1663,6 +1685,31 @@
                 }, I18n.format("multiplayer.texturePrompt.line1", new Object[0]), I18n.format("multiplayer.texturePrompt.line2", new Object[0]), 0));
             }
         }
+        // VIVE START - server told us that it has this mod too, allow extended reach, etc.
+        else if ("MC|ViveOK".equals(packetIn.func_149169_c()))
+        { //allowed, set to user preference.
+        	this.gameController.vrPlayer.noTeleportClient = false;
+            this.gameController.vrPlayer.setFreeMoveMode(Minecraft.getMinecraft().vrSettings.vrFreeMove);
+        } 
+        else if ("Vivecraft".equals(packetIn.func_149169_c()))
+        { 
+        	byte[] bytes = packetIn.func_149168_d();
+        	byte[] payload = ArrayUtils.remove(bytes, 0);
+        	PacketDiscriminators dis = PacketDiscriminators.values()[bytes[0]];
+        	switch (dis){
+        	case VERSION:
+            	this.gameController.vrPlayer.noTeleportClient = false;
+                this.gameController.vrPlayer.setFreeMoveMode(Minecraft.getMinecraft().vrSettings.vrFreeMove);
+                this.gameController.printChatMessage("Vivecraft server mod detected: " + new String(payload));
+        		break;
+        	case REQUESTDATA:
+        		NetworkHelper.serverWantsData = true;
+        	default:
+        		break;
+        	}
+
+        }
+        // VIVE END
     }
 
     /**
