--- a/net/minecraft/network/play/client/C03PacketPlayer.java
+++ b/net/minecraft/network/play/client/C03PacketPlayer.java
@@ -14,7 +14,7 @@
     protected double stance;
     protected float yaw;
     protected float pitch;
-    protected boolean field_149474_g;
+    protected boolean grounded;
     protected boolean field_149480_h;
     protected boolean rotating;
     private static final String __OBFID = "CL_00001360";
@@ -23,7 +23,7 @@
 
     public C03PacketPlayer(boolean p_i45256_1_)
     {
-        this.field_149474_g = p_i45256_1_;
+        this.grounded = p_i45256_1_;
     }
 
     /**
@@ -39,7 +39,7 @@
      */
     public void readPacketData(PacketBuffer data) throws IOException
     {
-        this.field_149474_g = data.readUnsignedByte() != 0;
+        this.grounded = data.readUnsignedByte() != 0;
     }
 
     /**
@@ -47,7 +47,7 @@
      */
     public void writePacketData(PacketBuffer data) throws IOException
     {
-        data.writeByte(this.field_149474_g ? 1 : 0);
+        data.writeByte(this.grounded ? 1 : 0);
     }
 
     public double getPositionX()
@@ -82,7 +82,7 @@
 
     public boolean func_149465_i()
     {
-        return this.field_149474_g;
+        return this.grounded;
     }
 
     public boolean func_149466_j()
@@ -123,7 +123,7 @@
             this.y = p_i45253_3_;
             this.stance = p_i45253_5_;
             this.z = p_i45253_7_;
-            this.field_149474_g = p_i45253_9_;
+            this.grounded = p_i45253_9_;
             this.field_149480_h = true;
         }
 
@@ -164,7 +164,7 @@
         {
             this.yaw = p_i45255_1_;
             this.pitch = p_i45255_2_;
-            this.field_149474_g = p_i45255_3_;
+            this.grounded = p_i45255_3_;
             this.rotating = true;
         }
 
@@ -206,7 +206,7 @@
             this.z = p_i45254_7_;
             this.yaw = p_i45254_9_;
             this.pitch = p_i45254_10_;
-            this.field_149474_g = p_i45254_11_;
+            this.grounded = p_i45254_11_;
             this.rotating = true;
             this.field_149480_h = true;
         }
