--- a/net/minecraft/entity/ai/EntityAIAvoidEntity.java
+++ b/net/minecraft/entity/ai/EntityAIAvoidEntity.java
@@ -30,12 +30,12 @@
     private Predicate field_179510_i;
     private static final String __OBFID = "CL_00001574";
 
-    public EntityAIAvoidEntity(EntityCreature p_i46384_1_, Class p_i46384_2_, float p_i46384_3_, double p_i46384_4_, double p_i46384_6_)
+    public EntityAIAvoidEntity(EntityCreature p_i46462_1_, Class p_i46462_2_, float p_i46462_3_, double p_i46462_4_, double p_i46462_6_)
     {
-        this(p_i46384_1_, p_i46384_2_, Predicates.alwaysTrue(), p_i46384_3_, p_i46384_4_, p_i46384_6_);
+        this(p_i46462_1_, p_i46462_2_, Predicates.alwaysTrue(), p_i46462_3_, p_i46462_4_, p_i46462_6_);
     }
 
-    public EntityAIAvoidEntity(EntityCreature p_i46385_1_, Class p_i46385_2_, Predicate p_i46385_3_, float p_i46385_4_, double p_i46385_5_, double p_i46385_7_)
+    public EntityAIAvoidEntity(EntityCreature p_i46463_1_, Class p_i46463_2_, Predicate p_i46463_3_, float p_i46463_4_, double p_i46463_5_, double p_i46463_7_)
     {
         this.field_179509_a = new Predicate()
         {
@@ -49,13 +49,13 @@
                 return this.func_180419_a((Entity)p_apply_1_);
             }
         };
-        this.theEntity = p_i46385_1_;
-        this.i = p_i46385_2_;
-        this.field_179510_i = p_i46385_3_;
-        this.field_179508_f = p_i46385_4_;
-        this.farSpeed = p_i46385_5_;
-        this.nearSpeed = p_i46385_7_;
-        this.entityPathNavigate = p_i46385_1_.getNavigator();
+        this.theEntity = p_i46463_1_;
+        this.i = p_i46463_2_;
+        this.field_179510_i = p_i46463_3_;
+        this.field_179508_f = p_i46463_4_;
+        this.farSpeed = p_i46463_5_;
+        this.nearSpeed = p_i46463_7_;
+        this.entityPathNavigate = p_i46463_1_.getNavigator();
         this.setMutexBits(1);
     }
 
