--- a/net/minecraft/client/multiplayer/PlayerControllerMP.java
+++ b/net/minecraft/client/multiplayer/PlayerControllerMP.java
@@ -25,6 +25,7 @@
 import net.minecraft.util.Vec3;
 import net.minecraft.world.World;
 import net.minecraft.world.WorldSettings;
+import net.minecraft.src.Reflector;
 
 public class PlayerControllerMP
 {
@@ -152,7 +153,15 @@
             {
                 var5.playAuxSFX(2001, x, y, z, Block.getIdFromBlock(var6) + (var5.getBlockMetadata(x, y, z) << 12));
                 int var7 = var5.getBlockMetadata(x, y, z);
-                boolean var8 = var5.setBlockToAir(x, y, z);
+                /** MINECRIFT FORGE **/
+                boolean var8;
+                if (Reflector.forgeExists()) {
+                    //var8 = var6.removedByPlayer(var5, mc.thePlayer, x, y, z);
+                    var8 = Reflector.callBoolean(var6, Reflector.Block_removedByPlayer, new Object[] {var5, mc.thePlayer, x, y, z});
+                } else {
+                    var8 = var5.setBlockToAir(x, y, z);
+                }
+                /** END MINECRIFT FORGE **/
 
                 if (var8)
                 {
@@ -300,6 +309,11 @@
             }
         }
     }
+ 
+    //JRBUDDA: This is the only thing this file does, necessary? Use reflection to modify field directly from elsewhere?
+    // VIVE START - function to allow damaging blocks immediately
+    public void clearBlockHitDelay() { this.blockHitDelay = 0; }
+    // VIVE END - function to allow damaging blocks immediately
 
     /**
      * player reach distance = 4F
@@ -401,7 +415,16 @@
         }
         else
         {
-            return itemStackIn.tryPlaceItemIntoWorld(player, worldIn, x, y, z, side, var9, var10, var11);
+            /** MINECRIFT FORGE **/
+            if (!itemStackIn.tryPlaceItemIntoWorld(player, worldIn, x, y, z, side, var9, var10, var11)) {
+                return false;
+            }
+            if (Reflector.forgeExists() && itemStackIn.stackSize <= 0) {
+                //MinecraftForge.EVENT_BUS.post(new PlayerDestroyItemEvent(player, itemStackIn));
+                Reflector.postForgeBusEvent(Reflector.ForgePlayerDestroyItemEvent_Constructor, player, itemStackIn);
+            }
+            return true;
+            /** END MINECRIFT FORGE **/
         }
     }
 
@@ -423,9 +446,15 @@
         {
             player.inventory.mainInventory[player.inventory.currentItem] = var5;
 
-            if (var5.stackSize == 0)
+            if (var5.stackSize <= 0)
             {
                 player.inventory.mainInventory[player.inventory.currentItem] = null;
+                /** MINECRIFT FORGE **/
+                if (Reflector.forgeExists()) {
+                    //MinecraftForge.EVENT_BUS.post(new PlayerDestroyItemEvent(player, var5));
+                    Reflector.postForgeBusEvent(Reflector.ForgePlayerDestroyItemEvent_Constructor, player, var5);
+                }
+                /** END MINECRIFT FORGE **/
             }
 
             return true;
