--- a/net/minecraft/client/multiplayer/PlayerControllerMP.java
+++ b/net/minecraft/client/multiplayer/PlayerControllerMP.java
@@ -111,6 +111,14 @@
     {
         this.currentGameType = p_78746_1_;
         this.currentGameType.configurePlayerCapabilities(this.mc.thePlayer.capabilities);
+		if (this.currentGameType == WorldSettings.GameType.SURVIVAL) {
+				mc.vrPlayer.vrMovementStyle.setStyle("Arc");
+				mc.vrSettings.allowPitchAffectsHeightWhileFlying = false;
+		} else {
+				//I don't know where else to put this  - JRBUDDA, ongamemodechanged?
+				mc.vrPlayer.vrMovementStyle.setStyle("Beam");
+				mc.vrSettings.allowPitchAffectsHeightWhileFlying = true;
+		}
     }
 
     /**
@@ -301,12 +309,22 @@
         }
     }
 
+    // VIVE START - function to allow damaging blocks immediately
+    public void clearBlockHitDelay() { this.blockHitDelay = 0; }
+    // VIVE END - function to allow damaging blocks immediately
+
     /**
      * player reach distance = 4F
      */
     public float getBlockReachDistance()
     {
+        // VIVE START - more reach in VR to reduce movement
+        if (!mc.vrPlayer.restrictedViveClient)
+        {
+            return 6.0F;
+        }
         return this.currentGameType.isCreative() ? 5.0F : 4.5F;
+        // VIVE END - more reach in VR to reduce movement
     }
 
     public void updateController()
