--- a/net/minecraft/client/gui/inventory/GuiContainerCreative.java
+++ b/net/minecraft/client/gui/inventory/GuiContainerCreative.java
@@ -4,6 +4,8 @@
 import java.util.Iterator;
 import java.util.List;
 import java.util.Map;
+
+import com.mtbs3d.minecrift.control.GuiScreenNavigator;
 import net.minecraft.client.gui.GuiButton;
 import net.minecraft.client.gui.GuiTextField;
 import net.minecraft.client.gui.achievement.GuiAchievements;
@@ -587,7 +589,9 @@
      */
     public void drawScreen(int p_73863_1_, int p_73863_2_, float p_73863_3_)
     {
-        boolean var4 = Mouse.isButtonDown(0);
+        /** MINECRIFT **/
+        boolean var4 = Mouse.isButtonDown(0) || GuiScreenNavigator.selectDepressed;
+        /** END MINECRIFT **/
         int var5 = this.field_147003_i;
         int var6 = this.field_147009_r;
         int var7 = var5 + 175;
