--- a/net/minecraft/client/renderer/culling/ClippingHelper.java
+++ b/net/minecraft/client/renderer/culling/ClippingHelper.java
@@ -7,7 +7,6 @@
     public float[] field_178626_c = new float[16];
     public float[] clippingMatrix = new float[16];
     private static final String __OBFID = "CL_00000977";
-    private static final String __OBFID = "CL_00000977";
 
     private float dot(float[] p_178624_1_, float p_178624_2_, float p_178624_4_, float p_178624_6_)
     {
