--- a/net/minecraft/client/settings/GameSettings.java
+++ b/net/minecraft/client/settings/GameSettings.java
@@ -683,7 +683,9 @@
         if (p_74306_1_ == GameSettings.Options.ENABLE_VSYNC)
         {
             this.enableVsync = !this.enableVsync;
-            Display.setVSyncEnabled(this.enableVsync);
+			/** MINECRIFT *  // Don't update Vsync - need to be on
+            Display.setVSyncEnabled(this.enableVsync); 
+			** END MINECRIFT */
         }
 
         if (p_74306_1_ == GameSettings.Options.USE_VBO)
@@ -2955,8 +2957,10 @@
 
     public void updateVSync()
     {
+		/** MINECRIFT * // Don't update - needs to be on
         Display.setVSyncEnabled(this.enableVsync);
-    }
+		** END MINECRIFT */ 
+	}
 
     private void updateWaterOpacity()
     {
