--- a/net/minecraft/client/settings/KeyBinding.java
+++ b/net/minecraft/client/settings/KeyBinding.java
@@ -10,17 +10,19 @@
 
 public class KeyBinding implements Comparable
 {
-    private static final List keybindArray = new ArrayList();
-    private static final IntHashMap hash = new IntHashMap();
-    private static final Set keybindSet = new HashSet();
-    private final String keyDescription;
-    private final int keyCodeDefault;
-    private final String keyCategory;
-    private int keyCode;
+    /** MINECRIFT */
+    public static final List keybindArray = new ArrayList();
+    public static final IntHashMap hash = new IntHashMap();
+    public static final Set keybindSet = new HashSet();
+    public final String keyDescription;
+    public final int keyCodeDefault;
+    public final String keyCategory;
+    public int keyCode;
 
-    /** because _303 wanted me to call it that(Caironater) */
-    private boolean pressed;
+//jrbudda changed all this around
     private int pressTime;
+    
+	/** END MINECRIFT */
     private static final String __OBFID = "CL_00000628";
 
     public static void onTick(int keyCode)
@@ -44,7 +46,7 @@
 
             if (var2 != null)
             {
-                var2.pressed = pressed;
+                var2.pressTime = 1;
             }
         }
     }
@@ -90,7 +92,7 @@
 
     public boolean getIsKeyPressed()
     {
-        return this.pressed;
+        return this.pressTime > 0;
     }
 
     public String getKeyCategory()
@@ -98,7 +100,11 @@
         return this.keyCategory;
     }
 
-    public boolean isPressed()
+    public void pressKey(){
+    	this.pressTime = 1;	
+    }
+    
+    public boolean consumePressed()
     {
         if (this.pressTime == 0)
         {
@@ -111,10 +117,9 @@
         }
     }
 
-    private void unpressKey()
+    public void unpressKey()
     {
         this.pressTime = 0;
-        this.pressed = false;
     }
 
     public String getKeyDescription()
