--- a/net/minecraft/client/settings/KeyBinding.java
+++ b/net/minecraft/client/settings/KeyBinding.java
@@ -8,6 +8,14 @@
 import net.minecraft.client.resources.I18n;
 import net.minecraft.util.IntHashMap;
 
+/**
+ * @author Administrator
+ *
+ */
+/**
+ * @author Administrator
+ *
+ */
 public class KeyBinding implements Comparable
 {
     private static final List keybindArray = new ArrayList();
@@ -35,7 +43,7 @@
             }
         }
     }
-
+      
     public static void setKeyBindState(int keyCode, boolean pressed)
     {
         if (keyCode != 0)
@@ -49,6 +57,24 @@
         }
     }
 
+    
+    //MINECRIFT NON STATIC METHODS
+    public static void setKeyBindState(KeyBinding kb, boolean pressed)
+    {
+            if (kb != null)
+            {
+                kb.pressed = pressed;
+                ++kb.pressTime;
+            }       
+    }
+    
+    public void pressKey(){
+    	setKeyBindState(this, true);
+    }
+
+    
+    //END MINECRIFT
+    
     public static void unPressAllKeys()
     {
         Iterator var0 = keybindArray.iterator();
@@ -59,6 +85,13 @@
             var1.unpressKey();
         }
     }
+    
+    public void unpressKey()
+    {
+        this.pressTime = 0;
+        this.pressed = false;
+    }
+    
 
     public static void resetKeyBindingArrayAndHash()
     {
@@ -98,6 +131,10 @@
         return this.keyCategory;
     }
 
+    
+    /**
+     * Return unconsumed preeTime and decrements presstime when called.
+     */
     public boolean isPressed()
     {
         if (this.pressTime == 0)
@@ -111,12 +148,6 @@
         }
     }
 
-    private void unpressKey()
-    {
-        this.pressTime = 0;
-        this.pressed = false;
-    }
-
     public String getKeyDescription()
     {
         return this.keyDescription;
