--- a/net/minecraft/client/renderer/ActiveRenderInfo.java
+++ b/net/minecraft/client/renderer/ActiveRenderInfo.java
@@ -4,6 +4,7 @@
 import java.nio.IntBuffer;
 import net.minecraft.block.Block;
 import net.minecraft.block.BlockLiquid;
+import net.minecraft.client.Minecraft;
 import net.minecraft.entity.EntityLivingBase;
 import net.minecraft.entity.player.EntityPlayer;
 import net.minecraft.util.MathHelper;
@@ -73,6 +74,10 @@
         int var4 = p_74583_1_ ? 1 : 0;
         float var5 = p_74583_0_.rotationPitch;
         float var6 = p_74583_0_.rotationYaw;
+        //VIVECRAFT
+        var5 = -Minecraft.getMinecraft().roomScale.getHMDPitch_World();
+        var6 = Minecraft.getMinecraft().roomScale.getHMDYaw_World();
+        //////
         rotationX = MathHelper.cos(var6 * (float)Math.PI / 180.0F) * (float)(1 - var4 * 2);
         rotationZ = MathHelper.sin(var6 * (float)Math.PI / 180.0F) * (float)(1 - var4 * 2);
         rotationYZ = -rotationZ * MathHelper.sin(var5 * (float)Math.PI / 180.0F) * (float)(1 - var4 * 2);
