--- a/net/minecraft/client/renderer/BlockFluidRenderer.java
+++ b/net/minecraft/client/renderer/BlockFluidRenderer.java
@@ -7,6 +7,7 @@
 import net.minecraft.client.renderer.texture.TextureAtlasSprite;
 import net.minecraft.client.renderer.texture.TextureMap;
 import net.minecraft.src.CustomColorizer;
+import net.minecraft.src.RenderEnv;
 import net.minecraft.util.BlockPos;
 import net.minecraft.util.EnumFacing;
 import net.minecraft.util.MathHelper;
@@ -17,7 +18,6 @@
     private TextureAtlasSprite[] field_178272_a = new TextureAtlasSprite[2];
     private TextureAtlasSprite[] field_178271_b = new TextureAtlasSprite[2];
     private static final String __OBFID = "CL_00002519";
-    private static final String __OBFID = "CL_00002519";
 
     public BlockFluidRenderer()
     {
@@ -44,7 +44,12 @@
         float var10 = (float)(var7 & 255) / 255.0F;
         boolean var11 = var5.shouldSideBeRendered(p_178270_1_, p_178270_3_.offsetUp(), EnumFacing.UP);
         boolean var12 = var5.shouldSideBeRendered(p_178270_1_, p_178270_3_.offsetDown(), EnumFacing.DOWN);
-        boolean[] var13 = new boolean[] {var5.shouldSideBeRendered(p_178270_1_, p_178270_3_.offsetNorth(), EnumFacing.NORTH), var5.shouldSideBeRendered(p_178270_1_, p_178270_3_.offsetSouth(), EnumFacing.SOUTH), var5.shouldSideBeRendered(p_178270_1_, p_178270_3_.offsetWest(), EnumFacing.WEST), var5.shouldSideBeRendered(p_178270_1_, p_178270_3_.offsetEast(), EnumFacing.EAST)};
+        RenderEnv renderEnv = RenderEnv.getInstance(p_178270_1_, p_178270_2_, p_178270_3_);
+        boolean[] var13 = renderEnv.getBorderFlags();
+        var13[0] = var5.shouldSideBeRendered(p_178270_1_, p_178270_3_.offsetNorth(), EnumFacing.NORTH);
+        var13[1] = var5.shouldSideBeRendered(p_178270_1_, p_178270_3_.offsetSouth(), EnumFacing.SOUTH);
+        var13[2] = var5.shouldSideBeRendered(p_178270_1_, p_178270_3_.offsetWest(), EnumFacing.WEST);
+        var13[3] = var5.shouldSideBeRendered(p_178270_1_, p_178270_3_.offsetEast(), EnumFacing.EAST);
 
         if (!var11 && !var12 && !var13[0] && !var13[1] && !var13[2] && !var13[3])
         {
@@ -150,67 +155,67 @@
                 var14 = true;
             }
 
-            for (int var561 = 0; var561 < 4; ++var561)
+            for (int var571 = 0; var571 < 4; ++var571)
             {
-                int var571 = 0;
                 int var581 = 0;
+                int var59 = 0;
 
-                if (var561 == 0)
+                if (var571 == 0)
                 {
-                    --var581;
+                    --var59;
                 }
 
-                if (var561 == 1)
+                if (var571 == 1)
                 {
-                    ++var581;
+                    ++var59;
                 }
 
-                if (var561 == 2)
+                if (var571 == 2)
                 {
-                    --var571;
+                    --var581;
                 }
 
-                if (var561 == 3)
+                if (var571 == 3)
                 {
-                    ++var571;
+                    ++var581;
                 }
 
-                BlockPos var59 = p_178270_3_.add(var571, 0, var581);
+                BlockPos var60 = p_178270_3_.add(var581, 0, var59);
                 var31 = var6[1];
 
-                if (var13[var561])
+                if (var13[var571])
                 {
                     double var56;
                     double var57;
                     double var58;
-                    double var60;
+                    double var61;
 
-                    if (var561 == 0)
+                    if (var571 == 0)
                     {
                         var36 = var20;
                         var37 = var23;
                         var56 = var24;
                         var58 = var24 + 1.0D;
                         var57 = var28 + (double)var30;
-                        var60 = var28 + (double)var30;
+                        var61 = var28 + (double)var30;
                     }
-                    else if (var561 == 1)
+                    else if (var571 == 1)
                     {
                         var36 = var22;
                         var37 = var21;
                         var56 = var24 + 1.0D;
                         var58 = var24;
                         var57 = var28 + 1.0D - (double)var30;
-                        var60 = var28 + 1.0D - (double)var30;
+                        var61 = var28 + 1.0D - (double)var30;
                     }
-                    else if (var561 == 2)
+                    else if (var571 == 2)
                     {
                         var36 = var21;
                         var37 = var20;
                         var56 = var24 + (double)var30;
                         var58 = var24 + (double)var30;
                         var57 = var28 + 1.0D;
-                        var60 = var28;
+                        var61 = var28;
                     }
                     else
                     {
@@ -219,7 +224,7 @@
                         var56 = var24 + 1.0D - (double)var30;
                         var58 = var24 + 1.0D - (double)var30;
                         var57 = var28;
-                        var60 = var28 + 1.0D;
+                        var61 = var28 + 1.0D;
                     }
 
                     var14 = true;
@@ -228,17 +233,17 @@
                     float var48 = var31.getInterpolatedV((double)((1.0F - var36) * 16.0F * 0.5F));
                     float var49 = var31.getInterpolatedV((double)((1.0F - var37) * 16.0F * 0.5F));
                     float var50 = var31.getInterpolatedV(8.0D);
-                    p_178270_4_.func_178963_b(var5.getMixedBrightnessForBlock(p_178270_1_, var59));
+                    p_178270_4_.func_178963_b(var5.getMixedBrightnessForBlock(p_178270_1_, var60));
                     float var51 = 1.0F;
-                    var51 *= var561 < 2 ? var17 : var18;
+                    var51 *= var571 < 2 ? var17 : var18;
                     p_178270_4_.func_178986_b(var16 * var51 * var8, var16 * var51 * var9, var16 * var51 * var10);
                     p_178270_4_.addVertexWithUV(var56, var26 + (double)var36, var57, (double)var46, (double)var48);
-                    p_178270_4_.addVertexWithUV(var58, var26 + (double)var37, var60, (double)var47, (double)var49);
-                    p_178270_4_.addVertexWithUV(var58, var26 + 0.0D, var60, (double)var47, (double)var50);
+                    p_178270_4_.addVertexWithUV(var58, var26 + (double)var37, var61, (double)var47, (double)var49);
+                    p_178270_4_.addVertexWithUV(var58, var26 + 0.0D, var61, (double)var47, (double)var50);
                     p_178270_4_.addVertexWithUV(var56, var26 + 0.0D, var57, (double)var46, (double)var50);
                     p_178270_4_.addVertexWithUV(var56, var26 + 0.0D, var57, (double)var46, (double)var50);
-                    p_178270_4_.addVertexWithUV(var58, var26 + 0.0D, var60, (double)var47, (double)var50);
-                    p_178270_4_.addVertexWithUV(var58, var26 + (double)var37, var60, (double)var47, (double)var49);
+                    p_178270_4_.addVertexWithUV(var58, var26 + 0.0D, var61, (double)var47, (double)var50);
+                    p_178270_4_.addVertexWithUV(var58, var26 + (double)var37, var61, (double)var47, (double)var49);
                     p_178270_4_.addVertexWithUV(var56, var26 + (double)var36, var57, (double)var46, (double)var48);
                 }
             }
