--- a/net/minecraft/client/renderer/RenderGlobal.java
+++ b/net/minecraft/client/renderer/RenderGlobal.java
@@ -20,6 +20,8 @@
 import java.util.Random;
 import java.util.Set;
 import java.util.concurrent.Callable;
+
+import com.mtbs3d.minecrift.settings.VRSettings;
 import net.minecraft.block.Block;
 import net.minecraft.block.BlockChest;
 import net.minecraft.block.BlockEnderChest;
@@ -32,6 +34,7 @@
 import net.minecraft.client.audio.PositionedSoundRecord;
 import net.minecraft.client.entity.EntityPlayerSP;
 import net.minecraft.client.gui.FontRenderer;
+import net.minecraft.client.model.ModelBase;
 import net.minecraft.client.multiplayer.WorldClient;
 import net.minecraft.client.particle.EntityFX;
 import net.minecraft.client.renderer.chunk.ChunkRenderDispatcher;
@@ -46,6 +49,7 @@
 import net.minecraft.client.renderer.culling.Frustrum;
 import net.minecraft.client.renderer.culling.ICamera;
 import net.minecraft.client.renderer.entity.RenderManager;
+import net.minecraft.client.renderer.entity.RenderPlayer;
 import net.minecraft.client.renderer.texture.TextureAtlasSprite;
 import net.minecraft.client.renderer.texture.TextureManager;
 import net.minecraft.client.renderer.texture.TextureMap;
@@ -711,7 +715,10 @@
             boolean oldFancyGraphics = this.mc.gameSettings.fancyGraphics;
             this.mc.gameSettings.fancyGraphics = Config.isDroppedItemsFancy();
             RenderGlobal.ContainerLocalRenderInformation var40;
-
+            /** MINECRIFT */
+			Entity thePlayer = null;
+            RenderPlayer renderPlayer = (RenderPlayer)this.mc.getRenderManager().getEntityRenderObject(this.mc.thePlayer);
+            /** END MINECRIFT */
             while (var391.hasNext())
             {
                 var40 = (RenderGlobal.ContainerLocalRenderInformation)var391.next();
@@ -732,20 +739,34 @@
 
                             if (var28 == this.mc.func_175606_aa() && this.mc.gameSettings.thirdPersonView == 0 && !var35 || var28.posY >= 0.0D && var28.posY < 256.0D && !this.theWorld.isBlockLoaded(new BlockPos(var28)))
                             {
-                                continue;
-                            }
+	                        /** MINECRIFT */
+	                            boolean skipRender = true;
 
-                            ++this.countEntitiesRendered;
+								if (this.mc.vrSettings.renderFullFirstPersonModelMode == VRSettings.RENDER_FIRST_PERSON_FULL)
+	                                skipRender = false;
 
+	                            if (skipRender)
+	                                continue;
+
+	                            thePlayer = var28;
+	                        }
+
+	                        //++this.countEntitiesRendered;
+	                        /** END MINECRIFT */
                             if (var28.getClass() == EntityItemFrame.class)
                             {
                                 var28.renderDistanceWeight = 0.06D;
                             }
-
-                            this.renderedEntity = var28;
-                            this.field_175010_j.renderEntitySimple(var28, partialTicks);
-                            this.renderedEntity = null;
-                        }
+	                        /** MINECRIFT */
+	                        if (var28 != thePlayer) // EXPERIMENTAL: Delay player render - see end of this function
+	                        {
+	                            ++this.countEntitiesRendered;
+	                            this.renderedEntity = var28;
+	                            this.field_175010_j.renderEntitySimple(var28, partialTicks);
+	                            this.renderedEntity = null;
+	                        }
+	                        /** END MINECRIFT */
+	                    }
 
                         if (!var31 && var28 instanceof EntityWitherSkull)
                         {
@@ -796,13 +817,8 @@
             {
                 Iterator var48 = this.n.iterator();
 
-                while (true)
+                while (var48.hasNext())
                 {
-                    if (!var48.hasNext())
-                    {
-                        break;
-                    }
-
                     TileEntity var39 = (TileEntity)var48.next();
 
                     if (forgeTileEntityPass)
@@ -888,6 +904,35 @@
                     TileEntityRendererDispatcher.instance.func_180546_a(var42, partialTicks, var47.getPartialBlockDamage());
                 }
             }
+			
+            /** MINECRIFT EXPERIMENTAL */
+            //OpenGLdebugging.dumpOpenGLstateToFileOnce("../Avatar.txt");
+            GlStateManager.blendFunc(GL11.GL_SRC_ALPHA, GL11.GL_ONE_MINUS_SRC_ALPHA);
+
+			// Render crosshair...
+            this.mc.entityRenderer.func_175072_h(); // Disable lightmap
+            this.mc.entityRenderer.renderCrosshairAtDepth();
+            this.mc.entityRenderer.func_180436_i(); // Enable lightmap
+
+            // ...and now render the player
+            if (thePlayer != null) 
+            {
+                ModelBase.vrMode = true;
+                ModelBase.vrOffset = this.mc.vrSettings.renderPlayerOffset;
+                ModelBase._renderHead = false;
+                ModelBase._renderHeadWear = this.mc.vrSettings.renderHeadWear;
+
+                ++this.countEntitiesRendered;
+                this.renderedEntity = thePlayer;
+                this.field_175010_j.renderEntitySimple(thePlayer, partialTicks);
+                this.renderedEntity = null;
+
+                ModelBase.vrMode = false;
+                ModelBase.vrOffset = 0f;
+                ModelBase._renderHead = true;
+                ModelBase._renderHeadWear = true;
+            }
+			/** END MINECRIFT EXPERIMENTAL */			
 
             this.func_174969_t();
             this.mc.entityRenderer.func_175072_h();
