--- a/net/minecraft/client/renderer/RenderBlocks.java
+++ b/net/minecraft/client/renderer/RenderBlocks.java
@@ -4916,7 +4916,6 @@
             var7 = var10;
             var8 = var11;
         }
-
         return Minecraft.isAmbientOcclusionEnabled() && blockType.getLightValue() == 0 ? (this.partialRenderBounds ? this.renderStandardBlockWithAmbientOcclusionPartial(blockType, blockX, blockY, blockZ, var6, var7, var8) : this.renderStandardBlockWithAmbientOcclusion(blockType, blockX, blockY, blockZ, var6, var7, var8)) : this.renderStandardBlockWithColorMultiplier(blockType, blockX, blockY, blockZ, var6, var7, var8);
     }
 
