--- a/net/minecraft/client/renderer/RenderGlobal.java
+++ b/net/minecraft/client/renderer/RenderGlobal.java
@@ -16,6 +16,8 @@
 import java.util.Random;
 import java.util.Set;
 import java.util.concurrent.Callable;
+
+import com.mtbs3d.minecrift.settings.VRSettings;
 import net.minecraft.block.Block;
 import net.minecraft.block.BlockChest;
 import net.minecraft.block.BlockEnderChest;
@@ -28,6 +30,7 @@
 import net.minecraft.client.audio.PositionedSoundRecord;
 import net.minecraft.client.entity.EntityPlayerSP;
 import net.minecraft.client.gui.FontRenderer;
+import net.minecraft.client.model.ModelBase;
 import net.minecraft.client.multiplayer.WorldClient;
 import net.minecraft.client.particle.EntityFX;
 import net.minecraft.client.renderer.chunk.ChunkRenderDispatcher;
@@ -42,6 +45,7 @@
 import net.minecraft.client.renderer.culling.Frustrum;
 import net.minecraft.client.renderer.culling.ICamera;
 import net.minecraft.client.renderer.entity.RenderManager;
+import net.minecraft.client.renderer.entity.RenderPlayer;
 import net.minecraft.client.renderer.texture.TextureAtlasSprite;
 import net.minecraft.client.renderer.texture.TextureManager;
 import net.minecraft.client.renderer.texture.TextureMap;
@@ -197,7 +201,6 @@
     public Entity renderedEntity;
     public Set chunksToResortTransparency = new LinkedHashSet();
     public Set chunksToUpdateForced = new LinkedHashSet();
-    private static final String __OBFID = "CL_00000954";
 
     public RenderGlobal(Minecraft mcIn)
     {
@@ -673,7 +676,10 @@
             boolean oldFancyGraphics = this.mc.gameSettings.fancyGraphics;
             this.mc.gameSettings.fancyGraphics = Config.isDroppedItemsFancy();
             RenderGlobal.ContainerLocalRenderInformation var371;
-
+            /** MINECRIFT */
+			Entity thePlayer = null;
+            RenderPlayer renderPlayer = (RenderPlayer)this.mc.getRenderManager().getEntityRenderObject(this.mc.thePlayer);
+            /** END MINECRIFT */
             while (var36.hasNext())
             {
                 var371 = (RenderGlobal.ContainerLocalRenderInformation)var36.next();
@@ -697,19 +703,33 @@
 
                         if (var31 == this.mc.func_175606_aa() && this.mc.gameSettings.thirdPersonView == 0 && !var39 || var31.posY >= 0.0D && var31.posY < 256.0D && !this.theWorld.isBlockLoaded(new BlockPos(var31)))
                         {
-                            continue;
-                        }
+                        /** MINECRIFT */
+                            boolean skipRender = true;
+
+							if (this.mc.vrSettings.renderFullFirstPersonModelMode == VRSettings.RENDER_FIRST_PERSON_FULL)
+                                skipRender = false;
 
-                        ++this.countEntitiesRendered;
+                            if (skipRender)
+                                continue;
 
+                            thePlayer = var31;
+                        }
+
+                        //++this.countEntitiesRendered;
+                        /** END MINECRIFT */
                         if (var31.getClass() == EntityItemFrame.class)
                         {
                             var31.renderDistanceWeight = 0.06D;
                         }
-
-                        this.renderedEntity = var31;
-                        this.field_175010_j.renderEntitySimple(var31, partialTicks);
-                        this.renderedEntity = null;
+                        /** MINECRIFT */
+                        if (var31 != thePlayer) // EXPERIMENTAL: Delay player render - see end of this function
+                        {
+                            ++this.countEntitiesRendered;
+                            this.renderedEntity = var31;
+                            this.field_175010_j.renderEntitySimple(var31, partialTicks);
+                            this.renderedEntity = null;
+                        }
+                        /** END MINECRIFT */
                     }
 
                     if (!var35 && var31 instanceof EntityWitherSkull)
@@ -801,6 +821,35 @@
                 }
             }
 
+            /** MINECRIFT EXPERIMENTAL */
+            //OpenGLdebugging.dumpOpenGLstateToFileOnce("../Avatar.txt");
+            GlStateManager.blendFunc(GL11.GL_SRC_ALPHA, GL11.GL_ONE_MINUS_SRC_ALPHA);
+
+			// Render crosshair...
+            this.mc.entityRenderer.func_175072_h(); // Disable lightmap
+            this.mc.entityRenderer.renderCrosshairAtDepth();
+            this.mc.entityRenderer.func_180436_i(); // Enable lightmap
+
+            // ...and now render the player
+            if (thePlayer != null) 
+            {
+                ModelBase.vrMode = true;
+                ModelBase.vrOffset = this.mc.vrSettings.renderPlayerOffset;
+                ModelBase._renderHead = false;
+                ModelBase._renderHeadWear = this.mc.vrSettings.renderHeadWear;
+
+                ++this.countEntitiesRendered;
+                this.renderedEntity = thePlayer;
+                this.field_175010_j.renderEntitySimple(thePlayer, partialTicks);
+                this.renderedEntity = null;
+
+                ModelBase.vrMode = false;
+                ModelBase.vrOffset = 0f;
+                ModelBase._renderHead = true;
+                ModelBase._renderHeadWear = true;
+            }
+			/** END MINECRIFT EXPERIMENTAL */
+
             this.func_174969_t();
             this.mc.entityRenderer.func_175072_h();
             this.mc.mcProfiler.endSection();
@@ -2330,15 +2379,10 @@
             var18.addCrashSectionCallable("Position", new Callable()
             {
                 private static final String __OBFID = "CL_00000955";
-                private static final String __OBFID = "CL_00000955";
                 public String call()
                 {
                     return CrashReportCategory.getCoordinateInfo(p_180442_3_, p_180442_5_, p_180442_7_);
                 }
-                public Object call() throws Exception
-                {
-                    return this.call();
-                }
             });
             throw new ReportedException(var17);
         }
@@ -2818,7 +2862,6 @@
         final Set field_178035_c;
         final int field_178032_d;
         private static final String __OBFID = "CL_00002534";
-        private static final String __OBFID = "CL_00002534";
 
         private ContainerLocalRenderInformation(RenderChunk p_i46248_2_, EnumFacing p_i46248_3_, int p_i46248_4_)
         {
@@ -2838,7 +2881,6 @@
     {
         static final int[] field_178037_a = new int[VertexFormatElement.EnumUseage.values().length];
         private static final String __OBFID = "CL_00002535";
-        private static final String __OBFID = "CL_00002535";
 
         static
         {
