--- a/net/minecraft/client/renderer/ThreadDownloadImageData.java
+++ b/net/minecraft/client/renderer/ThreadDownloadImageData.java
@@ -27,6 +27,7 @@
     private Thread imageThread;
     private boolean textureUploaded;
     private static final String __OBFID = "CL_00001049";
+    public Boolean imageFound = null;
 
     public ThreadDownloadImageData(File p_i1049_1_, String p_i1049_2_, ResourceLocation p_i1049_3_, IImageBuffer p_i1049_4_)
     {
@@ -38,18 +39,15 @@
 
     private void checkTextureUploaded()
     {
-        if (!this.textureUploaded)
+        if (!this.textureUploaded && this.bufferedImage != null)
         {
-            if (this.bufferedImage != null)
+            if (this.textureLocation != null)
             {
-                if (this.textureLocation != null)
-                {
-                    this.deleteGlTexture();
-                }
-
-                TextureUtil.uploadTextureImage(super.getGlTextureId(), this.bufferedImage);
-                this.textureUploaded = true;
+                this.deleteGlTexture();
             }
+
+            TextureUtil.uploadTextureImage(super.getGlTextureId(), this.bufferedImage);
+            this.textureUploaded = true;
         }
     }
 
@@ -67,6 +65,8 @@
         {
             this.imageBuffer.func_152634_a();
         }
+
+        this.imageFound = Boolean.valueOf(this.bufferedImage != null);
     }
 
     public void loadTexture(IResourceManager p_110551_1_) throws IOException
@@ -90,6 +90,8 @@
                     {
                         this.setBufferedImage(this.imageBuffer.parseUserSkin(this.bufferedImage));
                     }
+
+                    this.imageFound = Boolean.valueOf(this.bufferedImage != null);
                 }
                 catch (IOException var3)
                 {
@@ -121,32 +123,33 @@
                     var1.setDoOutput(false);
                     var1.connect();
 
-                    if (var1.getResponseCode() / 100 == 2)
+                    if (var1.getResponseCode() / 100 != 2)
                     {
-                        BufferedImage var2;
+                        return;
+                    }
 
-                        if (ThreadDownloadImageData.this.field_152434_e != null)
-                        {
-                            FileUtils.copyInputStreamToFile(var1.getInputStream(), ThreadDownloadImageData.this.field_152434_e);
-                            var2 = ImageIO.read(ThreadDownloadImageData.this.field_152434_e);
-                        }
-                        else
-                        {
-                            var2 = TextureUtil.func_177053_a(var1.getInputStream());
-                        }
-
-                        if (ThreadDownloadImageData.this.imageBuffer != null)
-                        {
-                            var2 = ThreadDownloadImageData.this.imageBuffer.parseUserSkin(var2);
-                        }
+                    BufferedImage var6;
 
-                        ThreadDownloadImageData.this.setBufferedImage(var2);
-                        return;
+                    if (ThreadDownloadImageData.this.field_152434_e != null)
+                    {
+                        FileUtils.copyInputStreamToFile(var1.getInputStream(), ThreadDownloadImageData.this.field_152434_e);
+                        var6 = ImageIO.read(ThreadDownloadImageData.this.field_152434_e);
+                    }
+                    else
+                    {
+                        var6 = TextureUtil.func_177053_a(var1.getInputStream());
+                    }
+
+                    if (ThreadDownloadImageData.this.imageBuffer != null)
+                    {
+                        var6 = ThreadDownloadImageData.this.imageBuffer.parseUserSkin(var6);
                     }
+
+                    ThreadDownloadImageData.this.setBufferedImage(var6);
                 }
-                catch (Exception var6)
+                catch (Exception var61)
                 {
-                    ThreadDownloadImageData.logger.error("Couldn\'t download http texture", var6);
+                    ThreadDownloadImageData.logger.error("Couldn\'t download http texture", var61);
                     return;
                 }
                 finally
@@ -155,6 +158,8 @@
                     {
                         var1.disconnect();
                     }
+
+                    ThreadDownloadImageData.this.imageFound = Boolean.valueOf(ThreadDownloadImageData.this.bufferedImage != null);
                 }
             }
         };
