--- a/net/minecraft/client/renderer/ThreadDownloadImageData.java
+++ b/net/minecraft/client/renderer/ThreadDownloadImageData.java
@@ -121,33 +121,33 @@
                     var1.setDoOutput(false);
                     var1.connect();
 
-                    if (var1.getResponseCode() / 100 == 2)
+                    if (var1.getResponseCode() / 100 != 2)
                     {
-                        BufferedImage var2;
+                        return;
+                    }
 
-                        if (ThreadDownloadImageData.this.field_152434_e != null)
-                        {
-                            FileUtils.copyInputStreamToFile(var1.getInputStream(), ThreadDownloadImageData.this.field_152434_e);
-                            var2 = ImageIO.read(ThreadDownloadImageData.this.field_152434_e);
-                        }
-                        else
-                        {
-                            var2 = ImageIO.read(var1.getInputStream());
-                        }
+                    BufferedImage var2;
 
-                        if (ThreadDownloadImageData.this.imageBuffer != null)
-                        {
-                            var2 = ThreadDownloadImageData.this.imageBuffer.parseUserSkin(var2);
-                        }
+                    if (ThreadDownloadImageData.this.field_152434_e != null)
+                    {
+                        FileUtils.copyInputStreamToFile(var1.getInputStream(), ThreadDownloadImageData.this.field_152434_e);
+                        var2 = ImageIO.read(ThreadDownloadImageData.this.field_152434_e);
+                    }
+                    else
+                    {
+                        var2 = ImageIO.read(var1.getInputStream());
+                    }
 
-                        ThreadDownloadImageData.this.func_147641_a(var2);
-                        return;
+                    if (ThreadDownloadImageData.this.imageBuffer != null)
+                    {
+                        var2 = ThreadDownloadImageData.this.imageBuffer.parseUserSkin(var2);
                     }
+
+                    ThreadDownloadImageData.this.func_147641_a(var2);
                 }
                 catch (Exception var6)
                 {
                     ThreadDownloadImageData.logger.error("Couldn\'t download http texture", var6);
-                    return;
                 }
                 finally
                 {
