--- a/net/minecraft/client/renderer/ThreadDownloadImageData.java
+++ b/net/minecraft/client/renderer/ThreadDownloadImageData.java
@@ -231,12 +231,11 @@
                 this.setBufferedImage(var2);
                 return;
             }
-
-            return;
         }
         catch (Exception var9)
         {
             logger.error("Couldn\'t download http texture: " + var9.getClass().getName() + ": " + var9.getMessage());
+            return;
         }
         finally
         {
