--- a/net/minecraft/world/GameRules.java
+++ b/net/minecraft/world/GameRules.java
@@ -9,7 +9,6 @@
 {
     private TreeMap theGameRules = new TreeMap();
     private static final String __OBFID = "CL_00000136";
-    private static final String __OBFID = "CL_00000136";
 
     public GameRules()
     {
@@ -138,7 +137,6 @@
         private double valueDouble;
         private final GameRules.ValueType type;
         private static final String __OBFID = "CL_00000137";
-        private static final String __OBFID = "CL_00000137";
 
         public Value(String value, GameRules.ValueType type)
         {
@@ -217,7 +215,6 @@
         private static final String __OBFID = "CL_00002151";
 
         private static final GameRules.ValueType[] $VALUES$ = new GameRules.ValueType[]{ANY_VALUE, BOOLEAN_VALUE, NUMERICAL_VALUE};
-        private static final String __OBFID = "CL_00002151";
 
         private ValueType(String p_i46405_1_, int p_i46405_2_, String p_i45750_1_, int p_i45750_2_) {}
     }
