--- a/net/minecraft/client/renderer/entity/RenderPlayer.java
+++ b/net/minecraft/client/renderer/entity/RenderPlayer.java
@@ -26,11 +26,27 @@
     private boolean field_177140_a;
     private static final String __OBFID = "CL_00001020";
 
+    /** MINECRIFT */
+    public float vrOffset;
+    public boolean _renderHeadWear;
+    public boolean _renderHead;
+    /** END MINECRIFT */
+
     public RenderPlayer(RenderManager p_i46102_1_)
     {
         this(p_i46102_1_, false);
     }
 
+    /** MINECRIFT */
+    public void setVR(boolean vrMode, boolean renderHeadWear, float offset) {
+
+        vrOffset = vrMode? offset : 0.0F;
+        this.mainModel.vrOffset = vrOffset;
+        this.mainModel._renderHead = _renderHead = !vrMode;
+        this.mainModel._renderHeadWear = _renderHeadWear = renderHeadWear;
+    }
+    /** END MINECRIFT */
+
     public RenderPlayer(RenderManager p_i46103_1_, boolean p_i46103_2_)
     {
         super(p_i46103_1_, new ModelPlayer(0.0F, p_i46103_2_), 0.5F);
@@ -68,17 +84,20 @@
     {
         ModelPlayer var2 = this.func_177136_g();
 
-        if (p_177137_1_.func_175149_v())
+        if (p_177137_1_.func_175149_v())          // Is spectator?
         {
-            var2.func_178719_a(false);
-            var2.bipedHead.showModel = true;
+            var2.func_178719_a(false);            // Set all body parts to show=false
+            var2.bipedHead.showModel = true;      // Re-enable head and headwear only
             var2.bipedHeadwear.showModel = true;
         }
         else
         {
             ItemStack var3 = p_177137_1_.inventory.getCurrentItem();
             var2.func_178719_a(true);
-            var2.bipedHeadwear.showModel = p_177137_1_.func_175148_a(EnumPlayerModelParts.HAT);
+            /** MINECRIFT */
+            var2.bipedHead.showModel = _renderHead ? true : false;
+            var2.bipedHeadwear.showModel = _renderHeadWear ? p_177137_1_.func_175148_a(EnumPlayerModelParts.HAT) : false;
+            /** MINECRIFT */
             var2.field_178730_v.showModel = p_177137_1_.func_175148_a(EnumPlayerModelParts.JACKET);
             var2.field_178733_c.showModel = p_177137_1_.func_175148_a(EnumPlayerModelParts.LEFT_PANTS_LEG);
             var2.field_178731_d.showModel = p_177137_1_.func_175148_a(EnumPlayerModelParts.RIGHT_PANTS_LEG);
