--- a/net/minecraft/client/renderer/entity/RenderManager.java
+++ b/net/minecraft/client/renderer/entity/RenderManager.java
@@ -3,7 +3,10 @@
 import java.util.HashMap;
 import java.util.Iterator;
 import java.util.Map;
+
+import de.fruitfly.ovr.enums.EyeType;
 import net.minecraft.block.Block;
+import net.minecraft.client.Minecraft;
 import net.minecraft.client.gui.FontRenderer;
 import net.minecraft.client.model.ModelChicken;
 import net.minecraft.client.model.ModelCow;
@@ -93,8 +96,9 @@
 
 public class RenderManager
 {
+	
     /** A map of entity classes and the associated renderer. */
-    private Map entityRenderMap = new HashMap();
+    public Map entityRenderMap = new HashMap(); ///SHADERSMOD make public
 
     /** The static instance of RenderManager. */
     public static RenderManager instance = new RenderManager();
@@ -233,8 +237,12 @@
         }
         else
         {
-            this.playerViewY = p_147938_4_.prevRotationYaw + (p_147938_4_.rotationYaw - p_147938_4_.prevRotationYaw) * p_147938_7_;
-            this.playerViewX = p_147938_4_.prevRotationPitch + (p_147938_4_.rotationPitch - p_147938_4_.prevRotationPitch) * p_147938_7_;
+            // VIVE START - use hmd facing, not player
+            //this.playerViewY = p_147938_4_.prevRotationYaw + (p_147938_4_.rotationYaw - p_147938_4_.prevRotationYaw) * p_147938_7_;
+            //this.playerViewX = p_147938_4_.prevRotationPitch + (p_147938_4_.rotationPitch - p_147938_4_.prevRotationPitch) * p_147938_7_;
+            this.playerViewY = -Minecraft.getMinecraft().roomScale.getHMDYaw_World();
+            this.playerViewX = -Minecraft.getMinecraft().roomScale.getHMDPitch_World();
+            // VIVE END - use hmd facing not player
         }
 
         if (p_147938_6_.thirdPersonView == 2)
