--- a/net/minecraft/client/renderer/entity/RenderFish.java
+++ b/net/minecraft/client/renderer/entity/RenderFish.java
@@ -1,5 +1,8 @@
 package net.minecraft.client.renderer.entity;
 
+import de.fruitfly.ovr.structs.Matrix4f;
+import de.fruitfly.ovr.structs.Vector3f;
+import jopenvr.OpenVRUtil;
 import net.minecraft.client.Minecraft;
 import net.minecraft.client.renderer.Tessellator;
 import net.minecraft.entity.Entity;
@@ -64,6 +67,17 @@
             double var27 = p_76986_1_.angler.prevPosZ + (p_76986_1_.angler.posZ - p_76986_1_.angler.prevPosZ) * (double)p_76986_9_ + var22.zCoord;
             double var29 = p_76986_1_.angler == Minecraft.getMinecraft().thePlayer ? 0.0D : (double)p_76986_1_.angler.getEyeHeight();
 
+            // VIVE START - use controller for fishing rod start point
+            if (p_76986_1_.angler == Minecraft.getMinecraft().thePlayer)
+            {
+                Vec3 aimSource = Minecraft.getMinecraft().roomScale.getControllerMainPos_World();
+                Vec3 aimRotation = Minecraft.getMinecraft().roomScale.getControllerMainDir_World();
+                var23= aimSource.xCoord + aimRotation.xCoord * 0.28f;
+                var25= aimSource.yCoord - aimRotation.yCoord * 0.28f;
+                var27 = aimSource.zCoord + aimRotation.zCoord * 0.28f;
+            }
+            // VIVE END - use controller for fishing rod start point
+
             if (this.renderManager.options.thirdPersonView > 0 || p_76986_1_.angler != Minecraft.getMinecraft().thePlayer)
             {
                 float var31 = (p_76986_1_.angler.prevRenderYawOffset + (p_76986_1_.angler.renderYawOffset - p_76986_1_.angler.prevRenderYawOffset) * p_76986_9_) * (float)Math.PI / 180.0F;
