--- a/net/minecraft/block/BlockOre.java
+++ b/net/minecraft/block/BlockOre.java
@@ -22,9 +22,9 @@
         this(Material.rock.getMaterialMapColor());
     }
 
-    public BlockOre(MapColor p_i46400_1_)
+    public BlockOre(MapColor p_i46451_1_)
     {
-        super(Material.rock, p_i46400_1_);
+        super(Material.rock, p_i46451_1_);
         this.setCreativeTab(CreativeTabs.tabBlock);
     }
 
