--- a/net/minecraft/block/BlockDirt.java
+++ b/net/minecraft/block/BlockDirt.java
@@ -119,17 +119,17 @@
         private static final BlockDirt.DirtType[] $VALUES = new BlockDirt.DirtType[]{DIRT, COARSE_DIRT, PODZOL};
         private static final String __OBFID = "CL_00002125";
 
-        private DirtType(String p_i46393_1_, int p_i46393_2_, int p_i46393_3_, String p_i46393_4_, MapColor p_i46393_5_)
+        private DirtType(String p_i46466_1_, int p_i46466_2_, int p_i46466_3_, String p_i46466_4_, MapColor p_i46466_5_)
         {
-            this(p_i46393_1_, p_i46393_2_, p_i46393_3_, p_i46393_4_, p_i46393_4_, p_i46393_5_);
+            this(p_i46466_1_, p_i46466_2_, p_i46466_3_, p_i46466_4_, p_i46466_4_, p_i46466_5_);
         }
 
-        private DirtType(String p_i46394_1_, int p_i46394_2_, int p_i46394_3_, String p_i46394_4_, String p_i46394_5_, MapColor p_i46394_6_)
+        private DirtType(String p_i46467_1_, int p_i46467_2_, int p_i46467_3_, String p_i46467_4_, String p_i46467_5_, MapColor p_i46467_6_)
         {
-            this.metadata = p_i46394_3_;
-            this.name = p_i46394_4_;
-            this.unlocalizedName = p_i46394_5_;
-            this.h = p_i46394_6_;
+            this.metadata = p_i46467_3_;
+            this.name = p_i46467_4_;
+            this.unlocalizedName = p_i46467_5_;
+            this.h = p_i46467_6_;
         }
 
         public int getMetadata()
