--- a/net/minecraft/block/BlockBasePressurePlate.java
+++ b/net/minecraft/block/BlockBasePressurePlate.java
@@ -21,9 +21,9 @@
         this(materialIn, materialIn.getMaterialMapColor());
     }
 
-    protected BlockBasePressurePlate(Material p_i46389_1_, MapColor p_i46389_2_)
+    protected BlockBasePressurePlate(Material p_i46442_1_, MapColor p_i46442_2_)
     {
-        super(p_i46389_1_, p_i46389_2_);
+        super(p_i46442_1_, p_i46442_2_);
         this.setCreativeTab(CreativeTabs.tabRedstone);
         this.setTickRandomly(true);
     }
