--- a/net/minecraft/block/BlockBreakable.java
+++ b/net/minecraft/block/BlockBreakable.java
@@ -18,10 +18,10 @@
         this(p_i45712_1_, p_i45712_2_, p_i45712_1_.getMaterialMapColor());
     }
 
-    protected BlockBreakable(Material p_i46397_1_, boolean p_i46397_2_, MapColor p_i46397_3_)
+    protected BlockBreakable(Material p_i46453_1_, boolean p_i46453_2_, MapColor p_i46453_3_)
     {
-        super(p_i46397_1_, p_i46397_3_);
-        this.ignoreSimilarity = p_i46397_2_;
+        super(p_i46453_1_, p_i46453_3_);
+        this.ignoreSimilarity = p_i46453_2_;
     }
 
     public boolean isOpaqueCube()
