--- a/net/minecraft/block/BlockStoneSlabNew.java
+++ b/net/minecraft/block/BlockStoneSlabNew.java
@@ -175,11 +175,11 @@
         private static final BlockStoneSlabNew.EnumType[] $VALUES = new BlockStoneSlabNew.EnumType[]{RED_SANDSTONE};
         private static final String __OBFID = "CL_00002086";
 
-        private EnumType(String p_i46399_1_, int p_i46399_2_, int p_i46399_3_, String p_i46399_4_, MapColor p_i46399_5_)
+        private EnumType(String p_i46458_1_, int p_i46458_2_, int p_i46458_3_, String p_i46458_4_, MapColor p_i46458_5_)
         {
-            this.field_176922_c = p_i46399_3_;
-            this.field_176919_d = p_i46399_4_;
-            this.e = p_i46399_5_;
+            this.field_176922_c = p_i46458_3_;
+            this.field_176919_d = p_i46458_4_;
+            this.e = p_i46458_5_;
         }
 
         public int func_176915_a()
