--- a/net/minecraft/block/BlockStoneSlab.java
+++ b/net/minecraft/block/BlockStoneSlab.java
@@ -178,17 +178,17 @@
         private static final BlockStoneSlab.EnumType[] $VALUES = new BlockStoneSlab.EnumType[]{STONE, SAND, WOOD, COBBLESTONE, BRICK, SMOOTHBRICK, NETHERBRICK, QUARTZ};
         private static final String __OBFID = "CL_00002056";
 
-        private EnumType(String p_i46407_1_, int p_i46407_2_, int p_i46407_3_, MapColor p_i46407_4_, String p_i46407_5_)
+        private EnumType(String p_i46456_1_, int p_i46456_2_, int p_i46456_3_, MapColor p_i46456_4_, String p_i46456_5_)
         {
-            this(p_i46407_1_, p_i46407_2_, p_i46407_3_, p_i46407_4_, p_i46407_5_, p_i46407_5_);
+            this(p_i46456_1_, p_i46456_2_, p_i46456_3_, p_i46456_4_, p_i46456_5_, p_i46456_5_);
         }
 
-        private EnumType(String p_i46408_1_, int p_i46408_2_, int p_i46408_3_, MapColor p_i46408_4_, String p_i46408_5_, String p_i46408_6_)
+        private EnumType(String p_i46457_1_, int p_i46457_2_, int p_i46457_3_, MapColor p_i46457_4_, String p_i46457_5_, String p_i46457_6_)
         {
-            this.field_176637_j = p_i46408_3_;
-            this.k = p_i46408_4_;
-            this.field_176638_k = p_i46408_5_;
-            this.field_176635_l = p_i46408_6_;
+            this.field_176637_j = p_i46457_3_;
+            this.k = p_i46457_4_;
+            this.field_176638_k = p_i46457_5_;
+            this.field_176635_l = p_i46457_6_;
         }
 
         public int func_176624_a()
