--- a/net/minecraft/block/BlockCompressedPowered.java
+++ b/net/minecraft/block/BlockCompressedPowered.java
@@ -11,9 +11,9 @@
 {
     private static final String __OBFID = "CL_00000287";
 
-    public BlockCompressedPowered(Material p_i46403_1_, MapColor p_i46403_2_)
+    public BlockCompressedPowered(Material p_i46447_1_, MapColor p_i46447_2_)
     {
-        super(p_i46403_1_, p_i46403_2_);
+        super(p_i46447_1_, p_i46447_2_);
     }
 
     /**
