--- a/net/minecraft/block/BlockContainer.java
+++ b/net/minecraft/block/BlockContainer.java
@@ -17,9 +17,9 @@
         this(materialIn, materialIn.getMaterialMapColor());
     }
 
-    protected BlockContainer(Material p_i46388_1_, MapColor p_i46388_2_)
+    protected BlockContainer(Material p_i46461_1_, MapColor p_i46461_2_)
     {
-        super(p_i46388_1_, p_i46388_2_);
+        super(p_i46461_1_, p_i46461_2_);
         this.isBlockContainer = true;
     }
 
