--- a/net/minecraft/block/BlockFenceGate.java
+++ b/net/minecraft/block/BlockFenceGate.java
@@ -22,9 +22,9 @@
     public static final PropertyBool field_176467_M = PropertyBool.create("in_wall");
     private static final String __OBFID = "CL_00000243";
 
-    public BlockFenceGate(BlockPlanks.EnumType p_i46396_1_)
+    public BlockFenceGate(BlockPlanks.EnumType p_i46432_1_)
     {
-        super(Material.wood, p_i46396_1_.c());
+        super(Material.wood, p_i46432_1_.c());
         this.setDefaultState(this.blockState.getBaseState().withProperty(field_176466_a, Boolean.valueOf(false)).withProperty(field_176465_b, Boolean.valueOf(false)).withProperty(field_176467_M, Boolean.valueOf(false)));
         this.setCreativeTab(CreativeTabs.tabRedstone);
     }
