--- a/net/minecraft/block/BlockBush.java
+++ b/net/minecraft/block/BlockBush.java
@@ -25,9 +25,9 @@
         this(materialIn, materialIn.getMaterialMapColor());
     }
 
-    protected BlockBush(Material p_i46391_1_, MapColor p_i46391_2_)
+    protected BlockBush(Material p_i46465_1_, MapColor p_i46465_2_)
     {
-        super(p_i46391_1_, p_i46391_2_);
+        super(p_i46465_1_, p_i46465_2_);
         this.setTickRandomly(true);
         float var3 = 0.2F;
         this.setBlockBounds(0.5F - var3, 0.0F, 0.5F - var3, 0.5F + var3, var3 * 3.0F, 0.5F + var3);
