--- a/net/minecraft/util/EnumFacing.java
+++ b/net/minecraft/util/EnumFacing.java
@@ -38,7 +38,6 @@
     private static final EnumFacing[] HORIZONTALS = new EnumFacing[4];
     private static final Map NAME_LOOKUP = Maps.newHashMap();
     private static final String __OBFID = "CL_00001201";
-    private static final String __OBFID = "CL_00001201";
 
     private EnumFacing(int indexIn, int oppositeIn, int horizontalIndexIn, String nameIn, EnumFacing.AxisDirection axisDirectionIn, EnumFacing.Axis axisIn, Vec3i directionVecIn)
     {
@@ -380,7 +379,6 @@
         private static final String __OBFID = "CL_00002321";
 
         private static final EnumFacing.Axis[] $VALUES$ = new EnumFacing.Axis[]{X, Y, Z};
-        private static final String __OBFID = "CL_00002321";
 
         private Axis(String p_i46398_1_, int p_i46398_2_, String p_i46015_1_, int p_i46015_2_, String name, EnumFacing.Plane plane)
         {
@@ -454,7 +452,6 @@
         private static final String __OBFID = "CL_00002320";
 
         private static final EnumFacing.AxisDirection[] $VALUES$ = new EnumFacing.AxisDirection[]{POSITIVE, NEGATIVE};
-        private static final String __OBFID = "CL_00002320";
 
         private AxisDirection(String p_i46380_1_, int p_i46380_2_, String p_i46014_1_, int p_i46014_2_, int offset, String description)
         {
@@ -480,7 +477,6 @@
         private static final String __OBFID = "CL_00002319";
 
         private static final EnumFacing.Plane[] $VALUES$ = new EnumFacing.Plane[]{HORIZONTAL, VERTICAL};
-        private static final String __OBFID = "CL_00002319";
 
         private Plane(String p_i46382_1_, int p_i46382_2_, String p_i46013_1_, int p_i46013_2_) {}
 
@@ -524,7 +520,6 @@
         static final int[] FACING_LOOKUP;
         static final int[] PLANE_LOOKUP = new int[EnumFacing.Plane.values().length];
         private static final String __OBFID = "CL_00002322";
-        private static final String __OBFID = "CL_00002322";
 
         static {
             try {
