--- a/net/minecraft/util/MovementInputFromOptions.java
+++ b/net/minecraft/util/MovementInputFromOptions.java
@@ -1,5 +1,7 @@
 package net.minecraft.util;
 
+import com.mtbs3d.minecrift.settings.VRSettings;
+import net.minecraft.client.Minecraft;
 import net.minecraft.client.settings.GameSettings;
 
 public class MovementInputFromOptions extends MovementInput
@@ -14,8 +16,12 @@
 
     public void updatePlayerMoveState()
     {
-        this.moveStrafe = 0.0F;
-        this.moveForward = 0.0F;
+        /** MINECRIFT */
+        VRSettings vr = Minecraft.getMinecraft().vrSettings;
+        this.moveStrafe = baseMoveStrafe;        // 0.0f;
+        this.moveForward = baseMoveForward;      // 0.0f;
+
+        /** END MINECRIFT */
 
         if (this.gameSettings.keyBindForward.getIsKeyPressed())
         {
@@ -45,5 +51,18 @@
             this.moveStrafe = (float)((double)this.moveStrafe * 0.3D);
             this.moveForward = (float)((double)this.moveForward * 0.3D);
         }
+
+        /** MINECRIFT */
+
+        this.moveForward = this.moveForward * vr.movementSpeedMultiplier;
+        this.moveStrafe  = this.moveStrafe * vr.movementSpeedMultiplier;
+
+        // Ensure no one component is greater > 1.3, sprint speed
+//        this.moveForward = Math.min(this.moveForward, 1.3f);
+//        this.moveStrafe  = Math.min(this.moveStrafe, 1.3f);
+
+        // Don't bother to normalise, vanilla doesn't
+
+        /** END MINECRIFT */
     }
 }
