--- a/net/minecraft/util/Vec3.java
+++ b/net/minecraft/util/Vec3.java
@@ -2,14 +2,16 @@
 
 public class Vec3
 {
+    /** MINECRIFT */
     /** X coordinate of Vec3D */
-    public final double xCoord;
+    public double xCoord;
 
     /** Y coordinate of Vec3D */
-    public final double yCoord;
+    public double yCoord;
 
     /** Z coordinate of Vec3D */
-    public final double zCoord;
+    public double zCoord;
+    /** END MINECRIFT */
     private static final String __OBFID = "CL_00000612";
 
     public Vec3(double x, double y, double z)
@@ -210,4 +212,52 @@
         double var8 = this.zCoord * (double)var2 - this.xCoord * (double)var3;
         return new Vec3(var4, var6, var8);
     }
+
+
+    /** MINECRIFT ADDITIONS BELOW */
+
+    /**
+     * Rotates the vector around the x axis by the specified angle.
+     */
+    public void rotateAroundX(float par1)
+    {
+        float var2 = MathHelper.cos(par1);
+        float var3 = MathHelper.sin(par1);
+        double var4 = this.xCoord;
+        double var6 = this.yCoord * (double)var2 + this.zCoord * (double)var3;
+        double var8 = this.zCoord * (double)var2 - this.yCoord * (double)var3;
+        this.xCoord = var4;
+        this.yCoord = var6;
+        this.zCoord = var8;
+    }
+
+    /**
+     * Rotates the vector around the y axis by the specified angle.
+     */
+    public void rotateAroundY(float par1)
+    {
+        float var2 = MathHelper.cos(par1);
+        float var3 = MathHelper.sin(par1);
+        double var4 = this.xCoord * (double)var2 + this.zCoord * (double)var3;
+        double var6 = this.yCoord;
+        double var8 = this.zCoord * (double)var2 - this.xCoord * (double)var3;
+        this.xCoord = var4;
+        this.yCoord = var6;
+        this.zCoord = var8;
+    }
+
+    /**
+     * Rotates the vector around the z axis by the specified angle.
+     */
+    public void rotateAroundZ(float par1)
+    {
+        float var2 = MathHelper.cos(par1);
+        float var3 = MathHelper.sin(par1);
+        double var4 = this.xCoord * (double)var2 + this.yCoord * (double)var3;
+        double var6 = this.yCoord * (double)var2 - this.xCoord * (double)var3;
+        double var8 = this.zCoord;
+        this.xCoord = var4;
+        this.yCoord = var6;
+        this.zCoord = var8;
+    }
 }
