--- a/net/minecraft/util/Vec3.java
+++ b/net/minecraft/util/Vec3.java
@@ -198,8 +198,8 @@
     }
 
     public String toString()
-    {
-        return "(" + this.xCoord + ", " + this.yCoord + ", " + this.zCoord + ")";
+    { //TODO put back.
+        return "(" + String.format("%.2f", this.xCoord) + ", " + String.format("%.2f", this.yCoord) + ", " + String.format("%.2f", this.zCoord)+ ")";
     }
 
     /**
