--- a/net/minecraft/profiler/Profiler.java
+++ b/net/minecraft/profiler/Profiler.java
@@ -103,6 +103,9 @@
         }
     }
 
+    private String last;
+    private String lastlast;
+    
     /**
      * End section
      */
@@ -112,6 +115,11 @@
         {
             if (this.profilingEnabled)
             {
+            	if(this.timestampList.isEmpty()) {
+            		System.out.print("No section to End!");
+            		return;
+            	}
+            	last = this.profilingSection.toString();
                 long var1 = System.nanoTime();
                 long var3 = ((Long)this.timestampList.remove(this.timestampList.size() - 1)).longValue();
                 this.sectionList.remove(this.sectionList.size() - 1);
