--- a/net/minecraft/client/model/ModelPlayer.java
+++ b/net/minecraft/client/model/ModelPlayer.java
@@ -14,7 +14,6 @@
     private ModelRenderer field_178736_x;
     private boolean field_178735_y;
     private static final String __OBFID = "CL_00002626";
-    private static final String __OBFID = "CL_00002626";
 
     public ModelPlayer(float p_i46304_1_, boolean p_i46304_2_)
     {
@@ -89,6 +88,8 @@
         }
         else
         {
+            /** MINECRIFT */
+            GlStateManager.translate(0.0F, 0.0F, vrOffset );
             if (p_78088_1_.isSneaking())
             {
                 GlStateManager.translate(0.0F, 0.2F, 0.0F);
@@ -99,6 +100,8 @@
             this.field_178734_a.render(p_78088_7_);
             this.field_178732_b.render(p_78088_7_);
             this.field_178730_v.render(p_78088_7_);
+            GlStateManager.translate(0.0F, 0.0F, -vrOffset );
+            /** END MINECRIFT */
         }
 
         GlStateManager.popMatrix();
