--- a/net/minecraft/client/model/ModelPlayer.java
+++ b/net/minecraft/client/model/ModelPlayer.java
@@ -88,6 +88,8 @@
         }
         else
         {
+            /** MINECRIFT */
+            GlStateManager.translate(0.0F, 0.0F, vrOffset );
             if (p_78088_1_.isSneaking())
             {
                 GlStateManager.translate(0.0F, 0.2F, 0.0F);
@@ -98,6 +100,8 @@
             this.field_178734_a.render(p_78088_7_);
             this.field_178732_b.render(p_78088_7_);
             this.field_178730_v.render(p_78088_7_);
+            GlStateManager.translate(0.0F, 0.0F, -vrOffset );
+            /** END MINECRIFT */
         }
 
         GlStateManager.popMatrix();
