--- a/net/minecraft/client/model/ModelBase.java
+++ b/net/minecraft/client/model/ModelBase.java
@@ -25,6 +25,12 @@
     public int textureHeight = 32;
     private static final String __OBFID = "CL_00000845";
 
+    /** MINECRIFT */
+    public boolean _renderHead = true;
+    public boolean _renderHeadWear = true;
+    public float vrOffset = 0.0F;
+    /** END MINECRIFT */
+
     /**
      * Sets the models various rotation angles then renders the model.
      */
