--- a/net/minecraft/client/model/ModelBase.java
+++ b/net/minecraft/client/model/ModelBase.java
@@ -13,6 +13,11 @@
     public float swingProgress;
     public boolean isRiding;
 
+    /** MINECRIFT **/
+    public boolean vrMode = false;
+    public boolean initialisedForVr = false;
+    /** END MINECRIFT **/
+
     /**
      * This is a list of all the boxes (ModelRenderer.class) in the current model.
      */
