--- a/net/minecraft/client/renderer/chunk/ChunkRenderWorker.java
+++ b/net/minecraft/client/renderer/chunk/ChunkRenderWorker.java
@@ -165,26 +165,28 @@
                 {
                     ChunkRenderWorker.this.func_178473_b(p_178474_1_);
                     p_178474_1_.func_178540_f().lock();
-
-                    try
+                    label52:
                     {
-                        if (p_178474_1_.func_178546_a() != ChunkCompileTaskGenerator.Status.UPLOADING)
+                        try
                         {
+                            if (p_178474_1_.func_178546_a() == ChunkCompileTaskGenerator.Status.UPLOADING)
+                            {
+                                p_178474_1_.func_178535_a(ChunkCompileTaskGenerator.Status.DONE);
+                                break label52;
+                            }
+
                             if (!p_178474_1_.func_178537_h())
                             {
                                 ChunkRenderWorker.LOGGER.warn("Chunk render task was " + p_178474_1_.func_178546_a() + " when I expected it to be uploading; aborting task");
                             }
-
-                            return;
+                        }
+                        finally
+                        {
+                            p_178474_1_.func_178540_f().unlock();
                         }
 
-                        p_178474_1_.func_178535_a(ChunkCompileTaskGenerator.Status.DONE);
+                        return;
                     }
-                    finally
-                    {
-                        p_178474_1_.func_178540_f().unlock();
-                    }
-
                     p_178474_1_.func_178536_b().func_178580_a(var7);
                 }
                 public void onFailure(Throwable p_onFailure_1_)
