--- a/net/minecraft/client/renderer/chunk/ChunkRenderWorker.java
+++ b/net/minecraft/client/renderer/chunk/ChunkRenderWorker.java
@@ -56,26 +56,28 @@
     protected void func_178474_a(final ChunkCompileTaskGenerator p_178474_1_) throws InterruptedException
     {
         p_178474_1_.func_178540_f().lock();
-
-        try
+        label235:
         {
-            if (p_178474_1_.func_178546_a() != ChunkCompileTaskGenerator.Status.PENDING)
+            try
             {
+                if (p_178474_1_.func_178546_a() == ChunkCompileTaskGenerator.Status.PENDING)
+                {
+                    p_178474_1_.func_178535_a(ChunkCompileTaskGenerator.Status.COMPILING);
+                    break label235;
+                }
+
                 if (!p_178474_1_.func_178537_h())
                 {
                     LOGGER.warn("Chunk render task was " + p_178474_1_.func_178546_a() + " when I expected it to be pending; ignoring task");
                 }
-
-                return;
+            }
+            finally
+            {
+                p_178474_1_.func_178540_f().unlock();
             }
 
-            p_178474_1_.func_178535_a(ChunkCompileTaskGenerator.Status.COMPILING);
+            return;
         }
-        finally
-        {
-            p_178474_1_.func_178540_f().unlock();
-        }
-
         Entity var2 = Minecraft.getMinecraft().func_175606_aa();
 
         if (var2 == null)
@@ -163,28 +165,26 @@
                 {
                     ChunkRenderWorker.this.func_178473_b(p_178474_1_);
                     p_178474_1_.func_178540_f().lock();
-                    label53:
+
+                    try
                     {
-                        try
+                        if (p_178474_1_.func_178546_a() != ChunkCompileTaskGenerator.Status.UPLOADING)
                         {
-                            if (p_178474_1_.func_178546_a() == ChunkCompileTaskGenerator.Status.UPLOADING)
-                            {
-                                p_178474_1_.func_178535_a(ChunkCompileTaskGenerator.Status.DONE);
-                                break label53;
-                            }
-
                             if (!p_178474_1_.func_178537_h())
                             {
                                 ChunkRenderWorker.LOGGER.warn("Chunk render task was " + p_178474_1_.func_178546_a() + " when I expected it to be uploading; aborting task");
                             }
-                        }
-                        finally
-                        {
-                            p_178474_1_.func_178540_f().unlock();
+
+                            return;
                         }
 
-                        return;
+                        p_178474_1_.func_178535_a(ChunkCompileTaskGenerator.Status.DONE);
                     }
+                    finally
+                    {
+                        p_178474_1_.func_178540_f().unlock();
+                    }
+
                     p_178474_1_.func_178536_b().func_178580_a(var7);
                 }
                 public void onFailure(Throwable p_onFailure_1_)
