--- a/net/minecraft/client/renderer/chunk/RenderChunk.java
+++ b/net/minecraft/client/renderer/chunk/RenderChunk.java
@@ -41,7 +41,6 @@
     private boolean field_178593_n;
     private static final String __OBFID = "CL_00002452";
     private BlockPos[] positionOffsets16;
-    private static final String __OBFID = "CL_00002452";
 
     public RenderChunk(World worldIn, RenderGlobal renderGlobalIn, BlockPos blockPosIn, int indexIn)
     {
@@ -132,6 +131,11 @@
                 return;
             }
 
+            if (this.field_178588_d == null)
+            {
+                return;
+            }
+
             var9 = new RegionRenderCache(this.field_178588_d, var7.add(-1, -1, -1), var8.add(1, 1, 1), 1);
             p_178581_4_.func_178543_a(var5);
         }
@@ -251,15 +255,16 @@
     public ChunkCompileTaskGenerator func_178582_e()
     {
         this.field_178587_g.lock();
-        ChunkCompileTaskGenerator var1;
+        ChunkCompileTaskGenerator var2;
 
         try
         {
+            ChunkCompileTaskGenerator var1;
+
             if (this.field_178599_i != null && this.field_178599_i.func_178546_a() == ChunkCompileTaskGenerator.Status.PENDING)
             {
                 var1 = null;
-                Object var2 = var1;
-                return (ChunkCompileTaskGenerator)var2;
+                return var1;
             }
 
             if (this.field_178599_i != null && this.field_178599_i.func_178546_a() != ChunkCompileTaskGenerator.Status.DONE)
@@ -271,13 +276,14 @@
             this.field_178599_i = new ChunkCompileTaskGenerator(this, ChunkCompileTaskGenerator.Type.RESORT_TRANSPARENCY);
             this.field_178599_i.func_178543_a(this.field_178590_b);
             var1 = this.field_178599_i;
+            var2 = var1;
         }
         finally
         {
             this.field_178587_g.unlock();
         }
 
-        return var1;
+        return var2;
     }
 
     private void func_178573_a(WorldRenderer p_178573_1_, BlockPos p_178573_2_)
