--- a/net/minecraft/util/AxisAlignedBB.java
+++ b/net/minecraft/util/AxisAlignedBB.java
@@ -43,7 +43,7 @@
     }
 
     /**
-     * Adds the coordinates to the bounding box extending it if the point lies outside the current ranges. Args: x, y, z
+     * Returns new BB, extending it if the point lies outside the current ranges. Args: x, y, z. 
      */
     public AxisAlignedBB addCoord(double x, double y, double z)
     {
@@ -88,7 +88,7 @@
     }
 
     /**
-     * Returns a bounding box expanded by the specified vector (if negative numbers are given it will shrink). Args: x,
+     * Returns a new bounding box expanded by the specified vector (if negative numbers are given it will shrink). Args: x,
      * y, z
      */
     public AxisAlignedBB expand(double x, double y, double z)
