--- a/net/minecraft/util/LongHashMap.java
+++ b/net/minecraft/util/LongHashMap.java
@@ -22,7 +22,6 @@
     /** count of times elements have been added/removed */
     private transient volatile int modCount;
     private static final String __OBFID = "CL_00001492";
-    private static final String __OBFID = "CL_00001492";
 
     public LongHashMap()
     {
@@ -255,7 +254,6 @@
         LongHashMap.Entry nextEntry;
         final int hash;
         private static final String __OBFID = "CL_00001493";
-        private static final String __OBFID = "CL_00001493";
 
         Entry(int p_i1553_1_, long p_i1553_2_, Object p_i1553_4_, LongHashMap.Entry p_i1553_5_)
         {
