--- a/net/minecraft/util/ObjectIntIdentityMap.java
+++ b/net/minecraft/util/ObjectIntIdentityMap.java
@@ -12,7 +12,6 @@
     private final IdentityHashMap field_148749_a = new IdentityHashMap(512);
     private final List field_148748_b = Lists.newArrayList();
     private static final String __OBFID = "CL_00001203";
-    private static final String __OBFID = "CL_00001203";
 
     public void put(Object key, int value)
     {
