--- a/net/minecraft/item/ItemBow.java
+++ b/net/minecraft/item/ItemBow.java
@@ -26,11 +26,11 @@
     /**
      * called when the player releases the use item button. Args: itemstack, world, entityplayer, itemInUseCount
      */
-    public void onPlayerStoppedUsing(ItemStack p_77615_1_, World p_77615_2_, EntityPlayer p_77615_3_, int p_77615_4_)
+    public void onPlayerStoppedUsing(ItemStack p_77615_1_, World world_, EntityPlayer player, int p_77615_4_)
     {
-        boolean var5 = p_77615_3_.capabilities.isCreativeMode || EnchantmentHelper.getEnchantmentLevel(Enchantment.infinity.effectId, p_77615_1_) > 0;
+        boolean var5 = player.capabilities.isCreativeMode || EnchantmentHelper.getEnchantmentLevel(Enchantment.infinity.effectId, p_77615_1_) > 0;
 
-        if (var5 || p_77615_3_.inventory.hasItem(Items.arrow))
+        if (var5 || player.inventory.hasItem(Items.arrow))
         {
             int var6 = this.getMaxItemUseDuration(p_77615_1_) - p_77615_4_;
             float var7 = (float)var6 / 20.0F;
@@ -46,7 +46,7 @@
                 var7 = 1.0F;
             }
 
-            EntityArrow var8 = new EntityArrow(p_77615_2_, p_77615_3_, var7 * 2.0F);
+            EntityArrow var8 = new EntityArrow(world_, player, var7 * 2.0F);
 
             if (var7 == 1.0F)
             {
@@ -72,8 +72,8 @@
                 var8.setFire(100);
             }
 
-            p_77615_1_.damageItem(1, p_77615_3_);
-            p_77615_2_.playSoundAtEntity(p_77615_3_, "random.bow", 1.0F, 1.0F / (itemRand.nextFloat() * 0.4F + 1.2F) + var7 * 0.5F);
+            p_77615_1_.damageItem(1, player);
+            world_.playSoundAtEntity(player, "random.bow", 1.0F, 1.0F / (itemRand.nextFloat() * 0.4F + 1.2F) + var7 * 0.5F);
 
             if (var5)
             {
@@ -81,12 +81,12 @@
             }
             else
             {
-                p_77615_3_.inventory.consumeInventoryItem(Items.arrow);
+                player.inventory.consumeInventoryItem(Items.arrow);
             }
 
-            if (!p_77615_2_.isRemote)
+            if (!world_.isRemote)
             {
-                p_77615_2_.spawnEntityInWorld(var8);
+                world_.spawnEntityInWorld(var8);
             }
         }
     }
