--- a/net/minecraft/client/entity/AbstractClientPlayer.java
+++ b/net/minecraft/client/entity/AbstractClientPlayer.java
@@ -30,7 +30,6 @@
     private NetworkPlayerInfo field_175157_a;
     private ResourceLocation ofLocationCape = null;
     private static final String __OBFID = "CL_00000935";
-    private static final String __OBFID = "CL_00000935";
 
     public AbstractClientPlayer(World worldIn, GameProfile p_i45074_2_)
     {
