--- a/net/minecraft/client/entity/EntityPlayerSP.java
+++ b/net/minecraft/client/entity/EntityPlayerSP.java
@@ -734,7 +734,9 @@
 
         boolean var1 = this.movementInput.jump;
         boolean var2 = this.movementInput.sneak;
-        float var3 = 0.8F;
+        /** MINECRIFT */
+        float var3 = 0.8F * this.mc.vrSettings.movementSpeedMultiplier;
+        /** END MINECRIFT */
         boolean var4 = this.movementInput.moveForward >= var3;
         this.movementInput.updatePlayerMoveState();
 
