--- a/net/minecraft/client/gui/GuiButton.java
+++ b/net/minecraft/client/gui/GuiButton.java
@@ -4,6 +4,7 @@
 import net.minecraft.client.audio.PositionedSoundRecord;
 import net.minecraft.client.audio.SoundHandler;
 import net.minecraft.client.renderer.OpenGlHelper;
+import net.minecraft.src.Reflector;
 import net.minecraft.util.ResourceLocation;
 import org.lwjgl.opengl.GL11;
 
@@ -34,6 +35,7 @@
     public boolean visible;
     protected boolean hovered;
     private static final String __OBFID = "CL_00000668";
+    public int packedFGColour;
 
     public GuiButton(int buttonId, int x, int y, String buttonText)
     {
@@ -84,6 +86,12 @@
             FontRenderer var4 = mc.fontRendererObj;
             mc.getTextureManager().bindTexture(buttonTextures);
             GL11.glColor4f(1.0F, 1.0F, 1.0F, 1.0F);
+			
+			/** MINECRIFT **/
+            if (mouseX > Minecraft.getMinecraft().displayWidth)
+                mouseX -= Minecraft.getMinecraft().displayWidth;
+			/** END MINECRIFT **/
+			
             this.hovered = mouseX >= this.xPosition && mouseY >= this.yPosition && mouseX < this.xPosition + this.width && mouseY < this.yPosition + this.height;
             int var5 = this.getHoverState(this.hovered);
             GL11.glEnable(GL11.GL_BLEND);
@@ -94,7 +102,11 @@
             this.mouseDragged(mc, mouseX, mouseY);
             int var6 = 14737632;
 
-            if (!this.enabled)
+            if (Reflector.forgeExists() && packedFGColour != 0)
+            {
+                var6 = packedFGColour;
+            }
+            else if (!this.enabled)
             {
                 var6 = 10526880;
             }
