--- a/net/minecraft/client/gui/GuiMainMenu.java
+++ b/net/minecraft/client/gui/GuiMainMenu.java
@@ -120,7 +120,7 @@
         this.updateCounter = rand.nextFloat();
         this.field_92025_p = "";
 
-        if (!GLContext.getCapabilities().OpenGL20 && !OpenGlHelper.func_153193_b())
+        if (!GLContext.getCapabilities().OpenGL20 && !OpenGlHelper.isShadersSupported())
         {
             this.field_92025_p = I18n.format("title.oldgl1", new Object[0]);
             this.field_146972_A = I18n.format("title.oldgl2", new Object[0]);
