--- a/net/minecraft/client/gui/GuiNewChat.java
+++ b/net/minecraft/client/gui/GuiNewChat.java
@@ -4,8 +4,11 @@
 import java.util.ArrayList;
 import java.util.Iterator;
 import java.util.List;
+
+import com.mtbs3d.minecrift.settings.VRSettings;
 import net.minecraft.client.Minecraft;
 import net.minecraft.entity.player.EntityPlayer;
+import net.minecraft.src.Reflector;
 import net.minecraft.util.ChatComponentText;
 import net.minecraft.util.EnumChatFormatting;
 import net.minecraft.util.IChatComponent;
@@ -51,10 +54,19 @@
                     var3 = true;
                 }
 
+                /** MINECRIFT **/
+                if (!Minecraft.getMinecraft().vrSettings.chatFadeAway)
+                    var3 = true;
                 float var7 = this.getChatScale();
-                int var8 = MathHelper.ceiling_float_int((float)this.getChatWidth() / var7);
+                int chatWidthPixels = this.getChatWidth() ;
+                int chatFullHeightPixels = calculateChatboxHeight( this.mc.gameSettings.chatHeightFocused );
+                int var8 = MathHelper.ceiling_float_int((float)chatWidthPixels/ var7);
+                ScaledResolution res = new ScaledResolution(this.mc, this.mc.displayWidth, this.mc.displayHeight);
+                int offsetX =  (int)(VRSettings.inst.chatOffsetX*(res.getScaledWidth()  - chatWidthPixels -2 ));
+                int offsetY = -(int)(VRSettings.inst.chatOffsetY*(res.getScaledHeight() - chatFullHeightPixels -20 ));
                 GL11.glPushMatrix();
-                GL11.glTranslatef(2.0F, 20.0F, 0.0F);
+                GL11.glTranslatef(2.0F + offsetX, 20 + offsetY, 0.0F);
+                /** END MINECRIFT **/
                 GL11.glScalef(var7, var7, 1.0F);
                 int var9;
                 int var11;
@@ -100,6 +112,11 @@
                                 byte var15 = 0;
                                 int var16 = -var9 * 9;
                                 drawRect(var15, var16 - 9, var15 + var8 + 4, var16, var14 / 2 << 24);
+                                /** MINECRIFT FORGE **/
+								if (Reflector.forgeExists()) {
+									GL11.glEnable(GL11.GL_BLEND); // FORGE: BugFix MC-36812 Chat Opacity Broken in 1.7.x
+								}
+								/** END MINECRIFT FORGE **/
                                 String var17 = var10.getChatComponent().getFormattedText();
                                 this.mc.fontRendererObj.drawStringWithShadow(var17, var15, var16 - 8, 16777215 + (var14 << 24));
                                 GL11.glDisable(GL11.GL_ALPHA_TEST);
