--- a/net/minecraft/client/gui/GuiOptions.java
+++ b/net/minecraft/client/gui/GuiOptions.java
@@ -1,5 +1,10 @@
 package net.minecraft.client.gui;
 
+import com.mtbs3d.minecrift.gui.framework.BaseGuiSettings;
+import com.mtbs3d.minecrift.gui.framework.GuiButtonEx;
+import com.mtbs3d.minecrift.gui.GuiMinecriftSettings;
+import com.mtbs3d.minecrift.settings.VRSettings;
+import net.minecraft.client.Minecraft;
 import net.minecraft.client.audio.PositionedSoundRecord;
 import net.minecraft.client.audio.SoundCategory;
 import net.minecraft.client.audio.SoundEventAccessorComposite;
@@ -9,8 +14,11 @@
 import net.minecraft.client.resources.I18n;
 import net.minecraft.client.settings.GameSettings;
 import net.minecraft.client.stream.IStream;
+import org.apache.commons.io.FilenameUtils;
 
-public class GuiOptions extends GuiScreen implements GuiYesNoCallback
+import java.io.File;
+
+public class GuiOptions extends /** MINECRIFT */ BaseGuiSettings /** END MINECRIFT */ implements GuiYesNoCallback
 {
     private static final GameSettings.Options[] field_146440_f = new GameSettings.Options[] {GameSettings.Options.FOV, GameSettings.Options.DIFFICULTY};
     private final GuiScreen field_146441_g;
@@ -22,6 +30,9 @@
 
     public GuiOptions(GuiScreen p_i1046_1_, GameSettings p_i1046_2_)
     {
+		/** MINECRIFT */
+        super( p_i1046_1_, Minecraft.getMinecraft().vrSettings ); 
+		/** END MINECRIFT */
         this.field_146441_g = p_i1046_1_;
         this.game_settings_1 = p_i1046_2_;
     }
@@ -42,7 +53,17 @@
 
             if (var5.getEnumFloat())
             {
-                this.buttonList.add(new GuiOptionSlider(var5.returnEnumOrdinal(), this.width / 2 - 155 + var1 % 2 * 160, this.height / 6 - 12 + 24 * (var1 >> 1), var5));
+                /** MINECRIFT */
+                if (var5.getEnumString().equalsIgnoreCase("options.fov"))
+                {
+                    GuiOptionButton var7 = new GuiOptionButton(666, this.width / 2 - 155 + var1 % 2 * 160, this.height / 6 - 12 + 24 * (var1 >> 1), "VR Settings...");
+                    this.buttonList.add(var7);
+                }
+                else
+                {
+	               	this.buttonList.add(new GuiOptionSlider(var5.returnEnumOrdinal(), this.width / 2 - 155 + var1 % 2 * 160, this.height / 6 - 12 + 24 * (var1 >> 1), var5));
+                }
+				/** END MINECRIFT */
             }
             else
             {
@@ -60,8 +81,10 @@
             ++var1;
         }
 
-        this.buttonList.add(new GuiButton(8675309, this.width / 2 + 5, this.height / 6 + 48 - 6, 150, 20, "Super Secret Settings...")
-        {
+        /** MINECRIFT */ // Be helpful and actually display the shader name
+        this.buttonList.add(new GuiButtonEx(8675309, this.width / 2 + 5, this.height / 6 + 48 - 6, 150, 20, /*"Super Secret Settings..."*/getShaderButtonName())
+        /** END MINECRIFT */
+		{
             private static final String __OBFID = "CL_00000701";
             public void playPressSound(SoundHandler soundHandlerIn)
             {
@@ -97,6 +120,9 @@
             if (button.id == 8675309)
             {
                 this.mc.entityRenderer.activateNextShader();
+				/** MINECRIFT **/
+                this.reinit = true;
+				/** END MINECRIFT **/
             }
 
             if (button.id == 101)
@@ -161,6 +187,14 @@
                     GuiStreamUnavailable.func_152321_a(this);
                 }
             }
+
+            /** MINECRIFT */
+            if (button.id == 666)
+            {
+                this.mc.gameSettings.saveOptions();
+                this.mc.displayGuiScreen(new GuiMinecriftSettings(this, this.mc.vrSettings, this.game_settings_1));
+            }
+			/** END MINECRIFT */
         }
     }
 
@@ -173,4 +207,15 @@
         this.drawCenteredString(this.fontRendererObj, this.field_146442_a, this.width / 2, 15, 16777215);
         super.drawScreen(mouseX, mouseY, partialTicks);
     }
+
+    /** MINECRIFT ADDITIONS BELOW */
+    public String getShaderButtonName()
+    {
+        String shaderFileName = (this.mc.vrSettings.shaderIndex != VRSettings.NO_SHADER ? this.mc.entityRenderer.shaderResourceLocations[this.mc.vrSettings.shaderIndex].toString() : "None");
+        System.out.println("ShaderIndex: " + this.mc.vrSettings.shaderIndex);
+        File shaderFile = new File(shaderFileName);
+        String shaderName = FilenameUtils.removeExtension(shaderFile.getName());
+        String shaderButtonName = "Shader: " + shaderName;
+        return shaderButtonName;
+    }
 }
