--- a/net/minecraft/client/gui/GuiOptions.java
+++ b/net/minecraft/client/gui/GuiOptions.java
@@ -1,6 +1,12 @@
 package net.minecraft.client.gui;
 
+import java.io.File;
 import java.io.IOException;
+
+import com.mtbs3d.minecrift.gui.BaseGuiSettings;
+import com.mtbs3d.minecrift.gui.GuiButtonEx;
+import com.mtbs3d.minecrift.gui.GuiMinecriftSettings;
+import net.minecraft.client.Minecraft;
 import net.minecraft.client.audio.PositionedSoundRecord;
 import net.minecraft.client.audio.SoundCategory;
 import net.minecraft.client.audio.SoundEventAccessorComposite;
@@ -13,8 +19,9 @@
 import net.minecraft.util.ChatComponentText;
 import net.minecraft.util.ChatComponentTranslation;
 import net.minecraft.world.EnumDifficulty;
+import org.apache.commons.io.FilenameUtils;
 
-public class GuiOptions extends GuiScreen implements GuiYesNoCallback
+public class GuiOptions extends /** MINECRIFT */BaseGuiSettings /** END MINECRIFT */ implements GuiYesNoCallback
 {
     private static final GameSettings.Options[] field_146440_f = new GameSettings.Options[] {GameSettings.Options.FOV};
     private final GuiScreen field_146441_g;
@@ -28,6 +35,9 @@
 
     public GuiOptions(GuiScreen p_i1046_1_, GameSettings p_i1046_2_)
     {
+		/** MINECRIFT */
+        super( p_i1046_1_, Minecraft.getMinecraft().vrSettings );
+		/** END MINECRIFT */
         this.field_146441_g = p_i1046_1_;
         this.game_settings_1 = p_i1046_2_;
     }
@@ -48,7 +58,17 @@
 
             if (var5.getEnumFloat())
             {
-                this.buttonList.add(new GuiOptionSlider(var5.returnEnumOrdinal(), this.width / 2 - 155 + var1 % 2 * 160, this.height / 6 - 12 + 24 * (var1 >> 1), var5));
+                /** MINECRIFT */
+                if (var5.getEnumString().equalsIgnoreCase("options.fov"))
+                {
+                    GuiOptionButton var7 = new GuiOptionButton(666, this.width / 2 - 155 + var1 % 2 * 160, this.height / 6 - 12 + 24 * (var1 >> 1), "VR Settings...");
+                    this.buttonList.add(var7);
+                }
+                else
+                {
+	               	this.buttonList.add(new GuiOptionSlider(var5.returnEnumOrdinal(), this.width / 2 - 155 + var1 % 2 * 160, this.height / 6 - 12 + 24 * (var1 >> 1), var5));
+                }
+				/** END MINECRIFT */
             }
             else
             {
@@ -81,7 +101,9 @@
         }
 
         this.buttonList.add(new GuiButton(110, this.width / 2 - 155, this.height / 6 + 48 - 6, 150, 20, I18n.format("options.skinCustomisation", new Object[0])));
-        this.buttonList.add(new GuiButton(8675309, this.width / 2 + 5, this.height / 6 + 48 - 6, 150, 20, "Super Secret Settings...")
+        /** MINECRIFT */ // Be helpful and actually display the shader name
+        this.buttonList.add(new GuiButtonEx(8675309, this.width / 2 + 5, this.height / 6 + 48 - 6, 150, 20, /*"Super Secret Settings..."*/getShaderButtonName())
+        /** END MINECRIFT */
         {
             private static final String __OBFID = "CL_00000701";
             public void playPressSound(SoundHandler soundHandlerIn)
@@ -222,6 +244,14 @@
                     GuiStreamUnavailable.func_152321_a(this);
                 }
             }
+
+            /** MINECRIFT */
+            if (button.id == 666)
+            {
+                this.mc.gameSettings.saveOptions();
+                this.mc.displayGuiScreen(new GuiMinecriftSettings(this, this.mc.vrSettings, this.game_settings_1));
+            }
+			/** END MINECRIFT */
         }
     }
 
@@ -234,4 +264,15 @@
         this.drawCenteredString(this.fontRendererObj, this.field_146442_a, this.width / 2, 15, 16777215);
         super.drawScreen(mouseX, mouseY, partialTicks);
     }
+
+    /** MINECRIFT ADDITIONS BELOW */
+    public String getShaderButtonName()
+    {
+        String shaderFileName = (this.mc.entityRenderer.shaderIndex != this.mc.entityRenderer.shaderCount ? this.mc.entityRenderer.shaderResourceLocations[this.mc.entityRenderer.shaderIndex].toString() : "None");
+        System.out.println("ShaderIndex: " + this.mc.entityRenderer.shaderIndex);
+        File shaderFile = new File(shaderFileName);
+        String shaderName = FilenameUtils.removeExtension(shaderFile.getName());
+        String shaderButtonName = "Shader: " + shaderName;
+        return shaderButtonName;
+    }
 }
