--- a/net/minecraft/client/gui/GuiSlot.java
+++ b/net/minecraft/client/gui/GuiSlot.java
@@ -1,5 +1,6 @@
 package net.minecraft.client.gui;
 
+import com.mtbs3d.minecrift.control.GuiScreenNavigator;
 import net.minecraft.client.Minecraft;
 import net.minecraft.client.renderer.GlStateManager;
 import net.minecraft.client.renderer.Tessellator;
@@ -329,7 +330,9 @@
     {
         if (this.isMouseYWithinSlotBounds(this.mouseY))
         {
-            if (Mouse.isButtonDown(0) && this.getEnabled())
+            /** MINECRIFT */   // Allow gui cursor emulation with clickable scroll bars
+            if (( Mouse.isButtonDown(0) || GuiScreenNavigator.selectDepressed )&& this.getEnabled())
+            /** END MINECRIFT */        
             {
                 if (this.initialClickY == -1.0F)
                 {
@@ -520,4 +523,18 @@
     {
         return this.slotHeight;
     }
+	
+	/** MINECRIFT ADDITIONS BELOW */
+
+    public int publicGetSize() { return getSize(); }
+    public int select( int item, boolean dblClick) {
+        amountScrolled = item * slotHeight;
+        bindAmountScrolled();
+        elementClicked(item, dblClick, mouseX, mouseY);
+
+        if( !dblClick ) {
+            return (int)(this.top + 4 - this.amountScrolled + (item+0.5) * this.slotHeight + this.headerPadding);
+        }
+        return 0;
+    };
 }
