--- a/net/minecraft/client/gui/GuiScreen.java
+++ b/net/minecraft/client/gui/GuiScreen.java
@@ -65,7 +65,9 @@
     public int height;
 
     /** A list of all the buttons in this container. */
-    protected List buttonList = Lists.newArrayList();
+	/** MINECRIFT */
+    public List buttonList = Lists.newArrayList();
+    /** END MINECRIFT */
 
     /** A list of all the labels in this container. */
     protected List labelList = Lists.newArrayList();
@@ -87,6 +89,11 @@
     private URI field_175286_t;
     private static final String __OBFID = "CL_00000710";
 
+	/** MINECRIFT */
+	public static int mouseOffsetX = 0;
+	public static int mouseOffsetY = 0;
+	/** END MINECRIFT */
+
     /**
      * Draws the screen and all the components in it. Args : mouseX, mouseY, renderPartialTicks
      */
@@ -569,9 +576,16 @@
      */
     public void handleMouseInput() throws IOException
     {
-        int var1 = Mouse.getEventX() * this.width / this.mc.displayWidth;
-        int var2 = this.height - Mouse.getEventY() * this.height / this.mc.displayHeight - 1;
+		/** MINECRIFT */
+
+        // Reset mouse offset
+        mouseOffsetX = -1;
+        mouseOffsetY = -1;
+
+        int var1 = getMouseX();
+        int var2 = getMouseY();
         int var3 = Mouse.getEventButton();
+		/** END MINECRIFT */
 
         if (Mouse.getEventButtonState())
         {
@@ -636,7 +650,16 @@
     {
         if (this.mc.theWorld != null)
         {
-            this.drawGradientRect(0, 0, this.width, this.height, -1072689136, -804253680);
+            /** MINECRIFT */
+            if (this.mc.vrSettings != null && this.mc.vrSettings.menuBackground == false)
+            {
+                this.drawGradientRect(0, 0, this.width, this.height, 0, 0);
+            }
+            else
+            {
+                this.drawGradientRect(0, 0, this.width, this.height, -1072689136, -804253680);
+            }
+            /** END MINECRIFT */
         }
         else
         {
@@ -746,4 +769,80 @@
     {
         this.setWorldAndResolution(mcIn, p_175273_2_, p_175273_3_);
     }
+    
+    /** MINECRIFT ADDITIONS BELOW */
+	
+    public void mouseDown( int rawX, int rawY, int button ) throws IOException
+    {
+        int var1 = rawX * this.width / this.mc.displayFBWidth;
+        int var2 = this.height - rawY * this.height / this.mc.displayFBHeight - 1;
+        this.eventButton = button;
+        this.lastMouseEvent = Minecraft.getSystemTime();
+        mouseClicked(var1, var2, button);
+    }
+    
+    public void mouseUp( int rawX, int rawY, int button )
+    {
+        int var1 = rawX * this.width / this.mc.displayFBWidth;
+        int var2 = this.height - rawY * this.height / this.mc.displayFBHeight - 1;
+        mouseReleased(var1, var2, button);
+    }
+    
+    public void mouseDrag( int rawX, int rawY )
+    {
+        int var1 = rawX * this.width / this.mc.displayFBWidth;
+        int var2 = this.height - rawY * this.height / this.mc.displayFBHeight - 1;
+        long var3 = Minecraft.getSystemTime() - this.lastMouseEvent;
+        this.mouseClickMove(var1, var2, this.eventButton, var3);
+    }
+
+    public void mouseGuiDown( int guiX, int guiY, int button ) throws IOException
+    {
+        this.eventButton = button;
+        this.lastMouseEvent = Minecraft.getSystemTime();
+        mouseClicked(guiX, guiY, button);
+    }
+    
+    public void mouseGuiUp( int guiX, int guiY, int button )
+    {
+        mouseReleased(guiX, guiY, button);
+    }
+    
+    public void mouseGuiDrag( int guiX, int guiY )
+    {
+        long var3 = Minecraft.getSystemTime() - this.lastMouseEvent;
+        this.mouseClickMove(guiX, guiY, this.eventButton, var3);
+    }
+
+    public int getMouseX()
+    {
+        int mX = Mouse.getX() * this.mc.currentScreen.width / this.mc.displayFBWidth;
+        mX += (this.mouseOffsetX == -1 ? 0 : this.mouseOffsetX);
+        return mX;
+    }
+
+    public int getMouseY()
+    {
+        int mY = this.mc.currentScreen.height - Mouse.getY() * this.mc.currentScreen.height / this.mc.displayFBHeight - 1;
+        mY += (this.mouseOffsetY == -1 ? 0 : this.mouseOffsetY);
+        return mY;
+    }
+
+    public int getActualMouseX()
+    {
+        int mX = Mouse.getX() * this.mc.currentScreen.width / this.mc.displayFBWidth;
+        return mX;
+    }
+
+    public int getActualMouseY()
+    {
+        int mY = this.mc.currentScreen.height - Mouse.getY() * this.mc.currentScreen.height / this.mc.displayFBHeight - 1;
+        return mY;
+    }
+
+    public void setResolution(int width, int height)
+    {
+        this.width = width;
+        this.height = height;
+    }
 }
