--- a/net/minecraft/client/gui/GuiTextField.java
+++ b/net/minecraft/client/gui/GuiTextField.java
@@ -1,9 +1,12 @@
 package net.minecraft.client.gui;
 
+import net.minecraft.client.Minecraft;
 import net.minecraft.client.renderer.Tessellator;
 import net.minecraft.util.ChatAllowedCharacters;
 import org.lwjgl.opengl.GL11;
 
+import com.mtbs3d.minecrift.provider.MCOpenVR;
+
 public class GuiTextField extends Gui
 {
     private final FontRenderer fontRendererInstance;
@@ -473,6 +476,13 @@
         if (this.canLoseFocus)
         {
             this.setFocused(var4);
+        }  else {
+        	
+           // VIVE START - show keyboard
+            ((MCOpenVR) Minecraft.getMinecraft().lookaimController).setKeyboardOverlayShowing(true, this);
+          // VIVE END - show keyboard
+          
+        	
         }
 
         if (this.isFocused && p_146192_3_ == 0)
@@ -486,7 +496,8 @@
 
             String var6 = this.fontRendererInstance.trimStringToWidth(this.text.substring(this.lineScrollOffset), this.getWidth());
             this.setCursorPosition(this.fontRendererInstance.trimStringToWidth(var6, var5).length() + this.lineScrollOffset);
-        }
+            
+         }
     }
 
     /**
@@ -673,6 +684,11 @@
         }
 
         this.isFocused = p_146195_1_;
+        
+        // VIVE START - show keyboard
+       if(p_146195_1_) ((MCOpenVR) Minecraft.getMinecraft().lookaimController).setKeyboardOverlayShowing(p_146195_1_, this);
+        // VIVE END - show keyboard
+        
     }
 
     /**
