--- a/net/minecraft/client/gui/GuiVideoSettings.java
+++ b/net/minecraft/client/gui/GuiVideoSettings.java
@@ -72,7 +72,12 @@
 
                 if (y.getEnumFloat())
                 {
-                    this.buttonList.add(new GuiOptionSlider(y.returnEnumOrdinal(), x, y1, y));
+                    /** MINECRIFT */ // disable frame rate settings (never VSync)
+                    GuiOptionSlider slider = new GuiOptionSlider(y.returnEnumOrdinal(), x, y1, y);
+                    if (y == GameSettings.Options.FRAMERATE_LIMIT)
+                        slider.enabled = false;
+                    this.buttonList.add(slider);
+    				/** END MINECRIFT */
                 }
                 else
                 {
@@ -192,7 +197,7 @@
         }
 
         this.drawString(this.fontRendererObj, ver, 2, this.height - 10, 8421504);
-        String verMc = "Minecraft 1.7.10";
+        String verMc = this.mc.MINECRIFT_VERSION;
         int lenMc = this.fontRendererObj.getStringWidth(verMc);
         this.drawString(this.fontRendererObj, verMc, this.width - lenMc - 2, this.height - 10, 8421504);
 
