--- a/net/minecraft/client/gui/GuiVideoSettings.java
+++ b/net/minecraft/client/gui/GuiVideoSettings.java
@@ -67,18 +67,23 @@
 
             if (y != null)
             {
-                x = this.width / 2 - 155 + var15 % 2 * 160;
-                int y1 = this.height / 6 + 21 * (var15 / 2) - 10;
+	            x = this.width / 2 - 155 + var15 % 2 * 160;
+	            int y1 = this.height / 6 + 21 * (var15 / 2) - 10;
 
-                if (y.getEnumFloat())
-                {
-                    this.buttonList.add(new GuiOptionSlider(y.returnEnumOrdinal(), x, y1, y));
-                }
-                else
-                {
-                    this.buttonList.add(new GuiOptionButton(y.returnEnumOrdinal(), x, y1, y, this.guiGameSettings.getKeyBinding(y)));
-                }
-            }
+	            if (y.getEnumFloat())
+	            {
+	                /** MINECRIFT */ // disable frame rate settings (always VSync) 
+	                GuiOptionSlider slider = new GuiOptionSlider(y.returnEnumOrdinal(), x, y1, y);
+	                if (y == GameSettings.Options.FRAMERATE_LIMIT)
+	                    slider.enabled = false;
+	                this.buttonList.add(slider);
+					/** END MINECRIFT */
+	            }
+	            else
+	            {
+	                this.buttonList.add(new GuiOptionButton(y.returnEnumOrdinal(), x, y1, y, this.guiGameSettings.getKeyBinding(y)));
+	            }
+			}
         }
 
         int var16 = this.height / 6 + 21 * (var15 / 2) - 10;
