--- a/net/minecraft/client/renderer/block/model/BakedQuad.java
+++ b/net/minecraft/client/renderer/block/model/BakedQuad.java
@@ -10,7 +10,6 @@
     protected final EnumFacing face;
     private static final String __OBFID = "CL_00002512";
     private TextureAtlasSprite sprite = null;
-    private static final String __OBFID = "CL_00002512";
 
     public BakedQuad(int[] p_i46232_1_, int p_i46232_2_, EnumFacing p_i46232_3_, TextureAtlasSprite sprite)
     {
