--- a/net/minecraft/client/renderer/block/model/BakedQuad.java
+++ b/net/minecraft/client/renderer/block/model/BakedQuad.java
@@ -10,7 +10,7 @@
     protected final EnumFacing face;
     private static final String __OBFID = "CL_00002512";
     private TextureAtlasSprite sprite = null;
-    private static final String __OBFID = "CL_00002512";
+    private int[] vertexDataSingle = null;
 
     public BakedQuad(int[] p_i46232_1_, int p_i46232_2_, EnumFacing p_i46232_3_, TextureAtlasSprite sprite)
     {
@@ -56,4 +56,41 @@
     {
         return this.face;
     }
+
+    public int[] getVertexDataSingle()
+    {
+        if (this.vertexDataSingle == null)
+        {
+            this.vertexDataSingle = makeVertexDataSingle(this.field_178215_a, this.sprite);
+        }
+
+        return this.vertexDataSingle;
+    }
+
+    private static int[] makeVertexDataSingle(int[] vd, TextureAtlasSprite sprite)
+    {
+        int[] vdSingle = new int[vd.length];
+        int ku;
+
+        for (ku = 0; ku < vdSingle.length; ++ku)
+        {
+            vdSingle[ku] = vd[ku];
+        }
+
+        ku = sprite.sheetWidth / sprite.getIconWidth();
+        int kv = sprite.sheetHeight / sprite.getIconHeight();
+
+        for (int i = 0; i < 4; ++i)
+        {
+            int pos = i * 7;
+            float tu = Float.intBitsToFloat(vdSingle[pos + 4]);
+            float tv = Float.intBitsToFloat(vdSingle[pos + 4 + 1]);
+            float u = sprite.toSingleU(tu);
+            float v = sprite.toSingleV(tv);
+            vdSingle[pos + 4] = Float.floatToRawIntBits(u);
+            vdSingle[pos + 4 + 1] = Float.floatToRawIntBits(v);
+        }
+
+        return vdSingle;
+    }
 }
