--- a/net/minecraft/client/renderer/block/model/BlockPart.java
+++ b/net/minecraft/client/renderer/block/model/BlockPart.java
@@ -26,13 +26,13 @@
     public final boolean field_178238_e;
     private static final String __OBFID = "CL_00002511";
 
-    public BlockPart(Vector3f p_i46379_1_, Vector3f p_i46379_2_, Map p_i46379_3_, BlockPartRotation p_i46379_4_, boolean p_i46379_5_)
+    public BlockPart(Vector3f p_i46446_1_, Vector3f p_i46446_2_, Map p_i46446_3_, BlockPartRotation p_i46446_4_, boolean p_i46446_5_)
     {
-        this.field_178241_a = p_i46379_1_;
-        this.field_178239_b = p_i46379_2_;
-        this.field_178240_c = p_i46379_3_;
-        this.field_178237_d = p_i46379_4_;
-        this.field_178238_e = p_i46379_5_;
+        this.field_178241_a = p_i46446_1_;
+        this.field_178239_b = p_i46446_2_;
+        this.field_178240_c = p_i46446_3_;
+        this.field_178237_d = p_i46446_4_;
+        this.field_178238_e = p_i46446_5_;
         this.func_178235_a();
     }
 
