--- a/net/minecraft/client/renderer/block/model/BlockPartRotation.java
+++ b/net/minecraft/client/renderer/block/model/BlockPartRotation.java
@@ -11,11 +11,11 @@
     public final boolean field_178341_d;
     private static final String __OBFID = "CL_00002506";
 
-    public BlockPartRotation(Vector3f p_i46380_1_, EnumFacing.Axis p_i46380_2_, float p_i46380_3_, boolean p_i46380_4_)
+    public BlockPartRotation(Vector3f p_i46438_1_, EnumFacing.Axis p_i46438_2_, float p_i46438_3_, boolean p_i46438_4_)
     {
-        this.field_178344_a = p_i46380_1_;
-        this.field_178342_b = p_i46380_2_;
-        this.field_178343_c = p_i46380_3_;
-        this.field_178341_d = p_i46380_4_;
+        this.field_178344_a = p_i46438_1_;
+        this.field_178342_b = p_i46438_2_;
+        this.field_178343_c = p_i46438_3_;
+        this.field_178341_d = p_i46438_4_;
     }
 }
