--- a/net/minecraft/client/renderer/block/model/ItemTransformVec3f.java
+++ b/net/minecraft/client/renderer/block/model/ItemTransformVec3f.java
@@ -19,11 +19,11 @@
     public final Vector3f field_178363_d;
     private static final String __OBFID = "CL_00002484";
 
-    public ItemTransformVec3f(Vector3f p_i46381_1_, Vector3f p_i46381_2_, Vector3f p_i46381_3_)
+    public ItemTransformVec3f(Vector3f p_i46459_1_, Vector3f p_i46459_2_, Vector3f p_i46459_3_)
     {
-        this.field_178364_b = new Vector3f(p_i46381_1_);
-        this.field_178365_c = new Vector3f(p_i46381_2_);
-        this.field_178363_d = new Vector3f(p_i46381_3_);
+        this.field_178364_b = new Vector3f(p_i46459_1_);
+        this.field_178365_c = new Vector3f(p_i46459_2_);
+        this.field_178363_d = new Vector3f(p_i46459_3_);
     }
 
     static class Deserializer implements JsonDeserializer
